// @aztec/bb.js/src/cbind/generated/api_types.ts
// AUTOGENERATED FILE - DO NOT EDIT

// Type aliases for primitive types
export type Field2 = [Uint8Array, Uint8Array];

// Public interfaces (exported)
export interface CircuitComputeVkResponse {
  bytes: Uint8Array;
  fields: Uint8Array[];
  hash: Uint8Array;
}

export interface CircuitInput {
  name: string;
  bytecode: Uint8Array;
  verificationKey: Uint8Array;
}

export interface ProofSystemSettings {
  ipaAccumulation: boolean;
  oracleHashType: string;
  disableZk: boolean;
  optimizedSolidityVerifier: boolean;
}

export interface CircuitProve {
  circuit: CircuitInput;
  witness: Uint8Array;
  settings: ProofSystemSettings;
}

export interface CircuitInputNoVK {
  name: string;
  bytecode: Uint8Array;
}

export interface CircuitComputeVk {
  circuit: CircuitInputNoVK;
  settings: ProofSystemSettings;
}

export interface CircuitStats {
  circuit: CircuitInput;
  includeGatesPerOpcode: boolean;
  settings: ProofSystemSettings;
}

export interface CircuitVerify {
  verificationKey: Uint8Array;
  publicInputs: Uint8Array[];
  proof: Uint8Array[];
  settings: ProofSystemSettings;
}

export interface ChonkComputeStandaloneVk {
  circuit: CircuitInputNoVK;
}

export interface ChonkComputeIvcVk {
  circuit: CircuitInputNoVK;
}

export interface ChonkStart {
  numCircuits: number;
}

export interface ChonkLoad {
  circuit: CircuitInput;
}

export interface ChonkAccumulate {
  witness: Uint8Array;
}

export interface ChonkProve {

}

export interface GoblinProof {
  mergeProof: Uint8Array[];
  eccvmProof: Uint8Array[];
  ipaProof: Uint8Array[];
  translatorProof: Uint8Array[];
}

export interface ChonkProof {
  megaProof: Uint8Array[];
  goblinProof: GoblinProof;
}

export interface ChonkVerify {
  proof: ChonkProof;
  vk: Uint8Array;
}

export interface VkAsFields {
  verificationKey: Uint8Array;
}

export interface MegaVkAsFields {
  verificationKey: Uint8Array;
}

export interface CircuitWriteSolidityVerifier {
  verificationKey: Uint8Array;
  settings: ProofSystemSettings;
}

export interface ChonkCheckPrecomputedVk {
  circuit: CircuitInput;
}

export interface ChonkStats {
  circuit: CircuitInputNoVK;
  includeGatesPerOpcode: boolean;
}

export interface Poseidon2Hash {
  inputs: Uint8Array[];
}

export interface Poseidon2Permutation {
  inputs: Uint8Array[];
}

export interface Poseidon2HashAccumulate {
  inputs: Uint8Array[];
}

export interface PedersenCommit {
  inputs: Uint8Array[];
  hashIndex: number;
}

export interface PedersenHash {
  inputs: Uint8Array[];
  hashIndex: number;
}

export interface PedersenHashBuffer {
  input: Uint8Array;
  hashIndex: number;
}

export interface Blake2s {
  data: Uint8Array;
}

export interface Blake2sToField {
  data: Uint8Array;
}

export interface AesEncrypt {
  plaintext: Uint8Array;
  iv: Uint8Array;
  key: Uint8Array;
  length: number;
}

export interface AesDecrypt {
  ciphertext: Uint8Array;
  iv: Uint8Array;
  key: Uint8Array;
  length: number;
}

export interface GrumpkinPoint {
  x: Uint8Array;
  y: Uint8Array;
}

export interface GrumpkinMul {
  point: GrumpkinPoint;
  scalar: Uint8Array;
}

export interface GrumpkinAdd {
  pointA: GrumpkinPoint;
  pointB: GrumpkinPoint;
}

export interface GrumpkinBatchMul {
  points: GrumpkinPoint[];
  scalar: Uint8Array;
}

export interface GrumpkinGetRandomFr {
  dummy: number;
}

export interface GrumpkinReduce512 {
  input: Uint8Array;
}

export interface Secp256k1Point {
  x: Uint8Array;
  y: Uint8Array;
}

export interface Secp256k1Mul {
  point: Secp256k1Point;
  scalar: Uint8Array;
}

export interface Secp256k1GetRandomFr {
  dummy: number;
}

export interface Secp256k1Reduce512 {
  input: Uint8Array;
}

export interface Bn254FrSqrt {
  input: Uint8Array;
}

export interface Bn254FqSqrt {
  input: Uint8Array;
}

export interface Bn254G1Point {
  x: Uint8Array;
  y: Uint8Array;
}

export interface Bn254G1Mul {
  point: Bn254G1Point;
  scalar: Uint8Array;
}

export interface Bn254G2Point {
  x: [Uint8Array, Uint8Array];
  y: [Uint8Array, Uint8Array];
}

export interface Bn254G2Mul {
  point: Bn254G2Point;
  scalar: Uint8Array;
}

export interface Bn254G1IsOnCurve {
  point: Bn254G1Point;
}

export interface Bn254G1FromCompressed {
  compressed: Uint8Array;
}

export interface SchnorrComputePublicKey {
  privateKey: Uint8Array;
}

export interface SchnorrConstructSignature {
  message: Uint8Array;
  privateKey: Uint8Array;
}

export interface SchnorrVerifySignature {
  message: Uint8Array;
  publicKey: GrumpkinPoint;
  s: Uint8Array;
  e: Uint8Array;
}

export interface EcdsaSecp256k1ComputePublicKey {
  privateKey: Uint8Array;
}

export interface EcdsaSecp256r1ComputePublicKey {
  privateKey: Uint8Array;
}

export interface EcdsaSecp256k1ConstructSignature {
  message: Uint8Array;
  privateKey: Uint8Array;
}

export interface EcdsaSecp256r1ConstructSignature {
  message: Uint8Array;
  privateKey: Uint8Array;
}

export interface EcdsaSecp256k1RecoverPublicKey {
  message: Uint8Array;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

export interface EcdsaSecp256r1RecoverPublicKey {
  message: Uint8Array;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

export interface EcdsaSecp256k1VerifySignature {
  message: Uint8Array;
  publicKey: Secp256k1Point;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

export interface Secp256r1Point {
  x: Uint8Array;
  y: Uint8Array;
}

export interface EcdsaSecp256r1VerifySignature {
  message: Uint8Array;
  publicKey: Secp256r1Point;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

export interface SrsInitSrs {
  pointsBuf: Uint8Array;
  numPoints: number;
  g2Point: Uint8Array;
}

export interface SrsInitGrumpkinSrs {
  pointsBuf: Uint8Array;
  numPoints: number;
}

export interface Shutdown {

}

export interface ErrorResponse {
  message: string;
}

export interface CircuitProveResponse {
  publicInputs: Uint8Array[];
  proof: Uint8Array[];
  vk: CircuitComputeVkResponse;
}

export interface CircuitInfoResponse {
  numGates: number;
  numGatesDyadic: number;
  numAcirOpcodes: number;
  gatesPerOpcode: number[];
}

export interface CircuitVerifyResponse {
  verified: boolean;
}

export interface ChonkComputeStandaloneVkResponse {
  bytes: Uint8Array;
  fields: Uint8Array[];
}

export interface ChonkComputeIvcVkResponse {
  bytes: Uint8Array;
}

export interface ChonkStartResponse {

}

export interface ChonkLoadResponse {

}

export interface ChonkAccumulateResponse {

}

export interface ChonkProveResponse {
  proof: ChonkProof;
}

export interface ChonkVerifyResponse {
  valid: boolean;
}

export interface VkAsFieldsResponse {
  fields: Uint8Array[];
}

export interface MegaVkAsFieldsResponse {
  fields: Uint8Array[];
}

export interface CircuitWriteSolidityVerifierResponse {
  solidityCode: string;
}

export interface ChonkCheckPrecomputedVkResponse {
  valid: boolean;
  actualVk: Uint8Array;
}

export interface ChonkStatsResponse {
  acirOpcodes: number;
  circuitSize: number;
  gatesPerOpcode: number[];
}

export interface Poseidon2HashResponse {
  hash: Uint8Array;
}

export interface Poseidon2PermutationResponse {
  outputs: Uint8Array[];
}

export interface Poseidon2HashAccumulateResponse {
  hash: Uint8Array;
}

export interface PedersenCommitResponse {
  point: GrumpkinPoint;
}

export interface PedersenHashResponse {
  hash: Uint8Array;
}

export interface PedersenHashBufferResponse {
  hash: Uint8Array;
}

export interface Blake2sResponse {
  hash: Uint8Array;
}

export interface Blake2sToFieldResponse {
  field: Uint8Array;
}

export interface AesEncryptResponse {
  ciphertext: Uint8Array;
}

export interface AesDecryptResponse {
  plaintext: Uint8Array;
}

export interface GrumpkinMulResponse {
  point: GrumpkinPoint;
}

export interface GrumpkinAddResponse {
  point: GrumpkinPoint;
}

export interface GrumpkinBatchMulResponse {
  points: GrumpkinPoint[];
}

export interface GrumpkinGetRandomFrResponse {
  value: Uint8Array;
}

export interface GrumpkinReduce512Response {
  value: Uint8Array;
}

export interface Secp256k1MulResponse {
  point: Secp256k1Point;
}

export interface Secp256k1GetRandomFrResponse {
  value: Uint8Array;
}

export interface Secp256k1Reduce512Response {
  value: Uint8Array;
}

export interface Bn254FrSqrtResponse {
  isSquareRoot: boolean;
  value: Uint8Array;
}

export interface Bn254FqSqrtResponse {
  isSquareRoot: boolean;
  value: Uint8Array;
}

export interface Bn254G1MulResponse {
  point: Bn254G1Point;
}

export interface Bn254G2MulResponse {
  point: Bn254G2Point;
}

export interface Bn254G1IsOnCurveResponse {
  isOnCurve: boolean;
}

export interface Bn254G1FromCompressedResponse {
  point: Bn254G1Point;
}

export interface SchnorrComputePublicKeyResponse {
  publicKey: GrumpkinPoint;
}

export interface SchnorrConstructSignatureResponse {
  s: Uint8Array;
  e: Uint8Array;
}

export interface SchnorrVerifySignatureResponse {
  verified: boolean;
}

export interface EcdsaSecp256k1ComputePublicKeyResponse {
  publicKey: Secp256k1Point;
}

export interface EcdsaSecp256r1ComputePublicKeyResponse {
  publicKey: Secp256r1Point;
}

export interface EcdsaSecp256k1ConstructSignatureResponse {
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

export interface EcdsaSecp256r1ConstructSignatureResponse {
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

export interface EcdsaSecp256k1RecoverPublicKeyResponse {
  publicKey: Secp256k1Point;
}

export interface EcdsaSecp256r1RecoverPublicKeyResponse {
  publicKey: Secp256r1Point;
}

export interface EcdsaSecp256k1VerifySignatureResponse {
  verified: boolean;
}

export interface EcdsaSecp256r1VerifySignatureResponse {
  verified: boolean;
}

export interface SrsInitSrsResponse {
  dummy: number;
}

export interface SrsInitGrumpkinSrsResponse {
  dummy: number;
}

export interface ShutdownResponse {

}

// Private Msgpack interfaces (not exported)
interface MsgpackCircuitComputeVkResponse {
  bytes: Uint8Array;
  fields: Uint8Array[];
  hash: Uint8Array;
}

interface MsgpackCircuitInput {
  name: string;
  bytecode: Uint8Array;
  verification_key: Uint8Array;
}

interface MsgpackProofSystemSettings {
  ipa_accumulation: boolean;
  oracle_hash_type: string;
  disable_zk: boolean;
  optimized_solidity_verifier: boolean;
}

interface MsgpackCircuitProve {
  circuit: MsgpackCircuitInput;
  witness: Uint8Array;
  settings: MsgpackProofSystemSettings;
}

interface MsgpackCircuitInputNoVK {
  name: string;
  bytecode: Uint8Array;
}

interface MsgpackCircuitComputeVk {
  circuit: MsgpackCircuitInputNoVK;
  settings: MsgpackProofSystemSettings;
}

interface MsgpackCircuitStats {
  circuit: MsgpackCircuitInput;
  include_gates_per_opcode: boolean;
  settings: MsgpackProofSystemSettings;
}

interface MsgpackCircuitVerify {
  verification_key: Uint8Array;
  public_inputs: Uint8Array[];
  proof: Uint8Array[];
  settings: MsgpackProofSystemSettings;
}

interface MsgpackChonkComputeStandaloneVk {
  circuit: MsgpackCircuitInputNoVK;
}

interface MsgpackChonkComputeIvcVk {
  circuit: MsgpackCircuitInputNoVK;
}

interface MsgpackChonkStart {
  num_circuits: number;
}

interface MsgpackChonkLoad {
  circuit: MsgpackCircuitInput;
}

interface MsgpackChonkAccumulate {
  witness: Uint8Array;
}

interface MsgpackChonkProve {

}

interface MsgpackGoblinProof {
  merge_proof: Uint8Array[];
  eccvm_proof: Uint8Array[];
  ipa_proof: Uint8Array[];
  translator_proof: Uint8Array[];
}

interface MsgpackChonkProof {
  mega_proof: Uint8Array[];
  goblin_proof: MsgpackGoblinProof;
}

interface MsgpackChonkVerify {
  proof: MsgpackChonkProof;
  vk: Uint8Array;
}

interface MsgpackVkAsFields {
  verification_key: Uint8Array;
}

interface MsgpackMegaVkAsFields {
  verification_key: Uint8Array;
}

interface MsgpackCircuitWriteSolidityVerifier {
  verification_key: Uint8Array;
  settings: MsgpackProofSystemSettings;
}

interface MsgpackChonkCheckPrecomputedVk {
  circuit: MsgpackCircuitInput;
}

interface MsgpackChonkStats {
  circuit: MsgpackCircuitInputNoVK;
  include_gates_per_opcode: boolean;
}

interface MsgpackPoseidon2Hash {
  inputs: Uint8Array[];
}

interface MsgpackPoseidon2Permutation {
  inputs: Uint8Array[];
}

interface MsgpackPoseidon2HashAccumulate {
  inputs: Uint8Array[];
}

interface MsgpackPedersenCommit {
  inputs: Uint8Array[];
  hash_index: number;
}

interface MsgpackPedersenHash {
  inputs: Uint8Array[];
  hash_index: number;
}

interface MsgpackPedersenHashBuffer {
  input: Uint8Array;
  hash_index: number;
}

interface MsgpackBlake2s {
  data: Uint8Array;
}

interface MsgpackBlake2sToField {
  data: Uint8Array;
}

interface MsgpackAesEncrypt {
  plaintext: Uint8Array;
  iv: Uint8Array;
  key: Uint8Array;
  length: number;
}

interface MsgpackAesDecrypt {
  ciphertext: Uint8Array;
  iv: Uint8Array;
  key: Uint8Array;
  length: number;
}

interface MsgpackGrumpkinPoint {
  x: Uint8Array;
  y: Uint8Array;
}

interface MsgpackGrumpkinMul {
  point: MsgpackGrumpkinPoint;
  scalar: Uint8Array;
}

interface MsgpackGrumpkinAdd {
  point_a: MsgpackGrumpkinPoint;
  point_b: MsgpackGrumpkinPoint;
}

interface MsgpackGrumpkinBatchMul {
  points: MsgpackGrumpkinPoint[];
  scalar: Uint8Array;
}

interface MsgpackGrumpkinGetRandomFr {
  dummy: number;
}

interface MsgpackGrumpkinReduce512 {
  input: Uint8Array;
}

interface MsgpackSecp256k1Point {
  x: Uint8Array;
  y: Uint8Array;
}

interface MsgpackSecp256k1Mul {
  point: MsgpackSecp256k1Point;
  scalar: Uint8Array;
}

interface MsgpackSecp256k1GetRandomFr {
  dummy: number;
}

interface MsgpackSecp256k1Reduce512 {
  input: Uint8Array;
}

interface MsgpackBn254FrSqrt {
  input: Uint8Array;
}

interface MsgpackBn254FqSqrt {
  input: Uint8Array;
}

interface MsgpackBn254G1Point {
  x: Uint8Array;
  y: Uint8Array;
}

interface MsgpackBn254G1Mul {
  point: MsgpackBn254G1Point;
  scalar: Uint8Array;
}

interface MsgpackBn254G2Point {
  x: [Uint8Array, Uint8Array];
  y: [Uint8Array, Uint8Array];
}

interface MsgpackBn254G2Mul {
  point: MsgpackBn254G2Point;
  scalar: Uint8Array;
}

interface MsgpackBn254G1IsOnCurve {
  point: MsgpackBn254G1Point;
}

interface MsgpackBn254G1FromCompressed {
  compressed: Uint8Array;
}

interface MsgpackSchnorrComputePublicKey {
  private_key: Uint8Array;
}

interface MsgpackSchnorrConstructSignature {
  message: Uint8Array;
  private_key: Uint8Array;
}

interface MsgpackSchnorrVerifySignature {
  message: Uint8Array;
  public_key: MsgpackGrumpkinPoint;
  s: Uint8Array;
  e: Uint8Array;
}

interface MsgpackEcdsaSecp256k1ComputePublicKey {
  private_key: Uint8Array;
}

interface MsgpackEcdsaSecp256r1ComputePublicKey {
  private_key: Uint8Array;
}

interface MsgpackEcdsaSecp256k1ConstructSignature {
  message: Uint8Array;
  private_key: Uint8Array;
}

interface MsgpackEcdsaSecp256r1ConstructSignature {
  message: Uint8Array;
  private_key: Uint8Array;
}

interface MsgpackEcdsaSecp256k1RecoverPublicKey {
  message: Uint8Array;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

interface MsgpackEcdsaSecp256r1RecoverPublicKey {
  message: Uint8Array;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

interface MsgpackEcdsaSecp256k1VerifySignature {
  message: Uint8Array;
  public_key: MsgpackSecp256k1Point;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

interface MsgpackSecp256r1Point {
  x: Uint8Array;
  y: Uint8Array;
}

interface MsgpackEcdsaSecp256r1VerifySignature {
  message: Uint8Array;
  public_key: MsgpackSecp256r1Point;
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

interface MsgpackSrsInitSrs {
  points_buf: Uint8Array;
  num_points: number;
  g2_point: Uint8Array;
}

interface MsgpackSrsInitGrumpkinSrs {
  points_buf: Uint8Array;
  num_points: number;
}

interface MsgpackShutdown {

}

interface MsgpackErrorResponse {
  message: string;
}

interface MsgpackCircuitProveResponse {
  public_inputs: Uint8Array[];
  proof: Uint8Array[];
  vk: MsgpackCircuitComputeVkResponse;
}

interface MsgpackCircuitInfoResponse {
  num_gates: number;
  num_gates_dyadic: number;
  num_acir_opcodes: number;
  gates_per_opcode: number[];
}

interface MsgpackCircuitVerifyResponse {
  verified: boolean;
}

interface MsgpackChonkComputeStandaloneVkResponse {
  bytes: Uint8Array;
  fields: Uint8Array[];
}

interface MsgpackChonkComputeIvcVkResponse {
  bytes: Uint8Array;
}

interface MsgpackChonkStartResponse {

}

interface MsgpackChonkLoadResponse {

}

interface MsgpackChonkAccumulateResponse {

}

interface MsgpackChonkProveResponse {
  proof: MsgpackChonkProof;
}

interface MsgpackChonkVerifyResponse {
  valid: boolean;
}

interface MsgpackVkAsFieldsResponse {
  fields: Uint8Array[];
}

interface MsgpackMegaVkAsFieldsResponse {
  fields: Uint8Array[];
}

interface MsgpackCircuitWriteSolidityVerifierResponse {
  solidity_code: string;
}

interface MsgpackChonkCheckPrecomputedVkResponse {
  valid: boolean;
  actual_vk: Uint8Array;
}

interface MsgpackChonkStatsResponse {
  acir_opcodes: number;
  circuit_size: number;
  gates_per_opcode: number[];
}

interface MsgpackPoseidon2HashResponse {
  hash: Uint8Array;
}

interface MsgpackPoseidon2PermutationResponse {
  outputs: Uint8Array[];
}

interface MsgpackPoseidon2HashAccumulateResponse {
  hash: Uint8Array;
}

interface MsgpackPedersenCommitResponse {
  point: MsgpackGrumpkinPoint;
}

interface MsgpackPedersenHashResponse {
  hash: Uint8Array;
}

interface MsgpackPedersenHashBufferResponse {
  hash: Uint8Array;
}

interface MsgpackBlake2sResponse {
  hash: Uint8Array;
}

interface MsgpackBlake2sToFieldResponse {
  field: Uint8Array;
}

interface MsgpackAesEncryptResponse {
  ciphertext: Uint8Array;
}

interface MsgpackAesDecryptResponse {
  plaintext: Uint8Array;
}

interface MsgpackGrumpkinMulResponse {
  point: MsgpackGrumpkinPoint;
}

interface MsgpackGrumpkinAddResponse {
  point: MsgpackGrumpkinPoint;
}

interface MsgpackGrumpkinBatchMulResponse {
  points: MsgpackGrumpkinPoint[];
}

interface MsgpackGrumpkinGetRandomFrResponse {
  value: Uint8Array;
}

interface MsgpackGrumpkinReduce512Response {
  value: Uint8Array;
}

interface MsgpackSecp256k1MulResponse {
  point: MsgpackSecp256k1Point;
}

interface MsgpackSecp256k1GetRandomFrResponse {
  value: Uint8Array;
}

interface MsgpackSecp256k1Reduce512Response {
  value: Uint8Array;
}

interface MsgpackBn254FrSqrtResponse {
  is_square_root: boolean;
  value: Uint8Array;
}

interface MsgpackBn254FqSqrtResponse {
  is_square_root: boolean;
  value: Uint8Array;
}

interface MsgpackBn254G1MulResponse {
  point: MsgpackBn254G1Point;
}

interface MsgpackBn254G2MulResponse {
  point: MsgpackBn254G2Point;
}

interface MsgpackBn254G1IsOnCurveResponse {
  is_on_curve: boolean;
}

interface MsgpackBn254G1FromCompressedResponse {
  point: MsgpackBn254G1Point;
}

interface MsgpackSchnorrComputePublicKeyResponse {
  public_key: MsgpackGrumpkinPoint;
}

interface MsgpackSchnorrConstructSignatureResponse {
  s: Uint8Array;
  e: Uint8Array;
}

interface MsgpackSchnorrVerifySignatureResponse {
  verified: boolean;
}

interface MsgpackEcdsaSecp256k1ComputePublicKeyResponse {
  public_key: MsgpackSecp256k1Point;
}

interface MsgpackEcdsaSecp256r1ComputePublicKeyResponse {
  public_key: MsgpackSecp256r1Point;
}

interface MsgpackEcdsaSecp256k1ConstructSignatureResponse {
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

interface MsgpackEcdsaSecp256r1ConstructSignatureResponse {
  r: Uint8Array;
  s: Uint8Array;
  v: number;
}

interface MsgpackEcdsaSecp256k1RecoverPublicKeyResponse {
  public_key: MsgpackSecp256k1Point;
}

interface MsgpackEcdsaSecp256r1RecoverPublicKeyResponse {
  public_key: MsgpackSecp256r1Point;
}

interface MsgpackEcdsaSecp256k1VerifySignatureResponse {
  verified: boolean;
}

interface MsgpackEcdsaSecp256r1VerifySignatureResponse {
  verified: boolean;
}

interface MsgpackSrsInitSrsResponse {
  dummy: number;
}

interface MsgpackSrsInitGrumpkinSrsResponse {
  dummy: number;
}

interface MsgpackShutdownResponse {

}

// Conversion functions (exported)
export function toCircuitComputeVkResponse(o: MsgpackCircuitComputeVkResponse): CircuitComputeVkResponse {
  if (o.bytes === undefined) { throw new Error("Expected bytes in CircuitComputeVkResponse deserialization"); }
  if (o.fields === undefined) { throw new Error("Expected fields in CircuitComputeVkResponse deserialization"); }
  if (o.hash === undefined) { throw new Error("Expected hash in CircuitComputeVkResponse deserialization"); };
  return {
    bytes: o.bytes,
    fields: o.fields,
    hash: o.hash,
  };
}

export function toCircuitInput(o: MsgpackCircuitInput): CircuitInput {
  if (o.name === undefined) { throw new Error("Expected name in CircuitInput deserialization"); }
  if (o.bytecode === undefined) { throw new Error("Expected bytecode in CircuitInput deserialization"); }
  if (o.verification_key === undefined) { throw new Error("Expected verification_key in CircuitInput deserialization"); };
  return {
    name: o.name,
    bytecode: o.bytecode,
    verificationKey: o.verification_key,
  };
}

export function toProofSystemSettings(o: MsgpackProofSystemSettings): ProofSystemSettings {
  if (o.ipa_accumulation === undefined) { throw new Error("Expected ipa_accumulation in ProofSystemSettings deserialization"); }
  if (o.oracle_hash_type === undefined) { throw new Error("Expected oracle_hash_type in ProofSystemSettings deserialization"); }
  if (o.disable_zk === undefined) { throw new Error("Expected disable_zk in ProofSystemSettings deserialization"); }
  if (o.optimized_solidity_verifier === undefined) { throw new Error("Expected optimized_solidity_verifier in ProofSystemSettings deserialization"); };
  return {
    ipaAccumulation: o.ipa_accumulation,
    oracleHashType: o.oracle_hash_type,
    disableZk: o.disable_zk,
    optimizedSolidityVerifier: o.optimized_solidity_verifier,
  };
}

export function toCircuitProve(o: MsgpackCircuitProve): CircuitProve {
  if (o.circuit === undefined) { throw new Error("Expected circuit in CircuitProve deserialization"); }
  if (o.witness === undefined) { throw new Error("Expected witness in CircuitProve deserialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitProve deserialization"); };
  return {
    circuit: toCircuitInput(o.circuit),
    witness: o.witness,
    settings: toProofSystemSettings(o.settings),
  };
}

export function toCircuitInputNoVK(o: MsgpackCircuitInputNoVK): CircuitInputNoVK {
  if (o.name === undefined) { throw new Error("Expected name in CircuitInputNoVK deserialization"); }
  if (o.bytecode === undefined) { throw new Error("Expected bytecode in CircuitInputNoVK deserialization"); };
  return {
    name: o.name,
    bytecode: o.bytecode,
  };
}

export function toCircuitComputeVk(o: MsgpackCircuitComputeVk): CircuitComputeVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in CircuitComputeVk deserialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitComputeVk deserialization"); };
  return {
    circuit: toCircuitInputNoVK(o.circuit),
    settings: toProofSystemSettings(o.settings),
  };
}

export function toCircuitStats(o: MsgpackCircuitStats): CircuitStats {
  if (o.circuit === undefined) { throw new Error("Expected circuit in CircuitStats deserialization"); }
  if (o.include_gates_per_opcode === undefined) { throw new Error("Expected include_gates_per_opcode in CircuitStats deserialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitStats deserialization"); };
  return {
    circuit: toCircuitInput(o.circuit),
    includeGatesPerOpcode: o.include_gates_per_opcode,
    settings: toProofSystemSettings(o.settings),
  };
}

export function toCircuitVerify(o: MsgpackCircuitVerify): CircuitVerify {
  if (o.verification_key === undefined) { throw new Error("Expected verification_key in CircuitVerify deserialization"); }
  if (o.public_inputs === undefined) { throw new Error("Expected public_inputs in CircuitVerify deserialization"); }
  if (o.proof === undefined) { throw new Error("Expected proof in CircuitVerify deserialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitVerify deserialization"); };
  return {
    verificationKey: o.verification_key,
    publicInputs: o.public_inputs,
    proof: o.proof,
    settings: toProofSystemSettings(o.settings),
  };
}

export function toChonkComputeStandaloneVk(o: MsgpackChonkComputeStandaloneVk): ChonkComputeStandaloneVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkComputeStandaloneVk deserialization"); };
  return {
    circuit: toCircuitInputNoVK(o.circuit),
  };
}

export function toChonkComputeIvcVk(o: MsgpackChonkComputeIvcVk): ChonkComputeIvcVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkComputeIvcVk deserialization"); };
  return {
    circuit: toCircuitInputNoVK(o.circuit),
  };
}

export function toChonkStart(o: MsgpackChonkStart): ChonkStart {
  if (o.num_circuits === undefined) { throw new Error("Expected num_circuits in ChonkStart deserialization"); };
  return {
    numCircuits: o.num_circuits,
  };
}

export function toChonkLoad(o: MsgpackChonkLoad): ChonkLoad {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkLoad deserialization"); };
  return {
    circuit: toCircuitInput(o.circuit),
  };
}

export function toChonkAccumulate(o: MsgpackChonkAccumulate): ChonkAccumulate {
  if (o.witness === undefined) { throw new Error("Expected witness in ChonkAccumulate deserialization"); };
  return {
    witness: o.witness,
  };
}

export function toChonkProve(o: MsgpackChonkProve): ChonkProve {
  return {};
}

export function toGoblinProof(o: MsgpackGoblinProof): GoblinProof {
  if (o.merge_proof === undefined) { throw new Error("Expected merge_proof in GoblinProof deserialization"); }
  if (o.eccvm_proof === undefined) { throw new Error("Expected eccvm_proof in GoblinProof deserialization"); }
  if (o.ipa_proof === undefined) { throw new Error("Expected ipa_proof in GoblinProof deserialization"); }
  if (o.translator_proof === undefined) { throw new Error("Expected translator_proof in GoblinProof deserialization"); };
  return {
    mergeProof: o.merge_proof,
    eccvmProof: o.eccvm_proof,
    ipaProof: o.ipa_proof,
    translatorProof: o.translator_proof,
  };
}

export function toChonkProof(o: MsgpackChonkProof): ChonkProof {
  if (o.mega_proof === undefined) { throw new Error("Expected mega_proof in ChonkProof deserialization"); }
  if (o.goblin_proof === undefined) { throw new Error("Expected goblin_proof in ChonkProof deserialization"); };
  return {
    megaProof: o.mega_proof,
    goblinProof: toGoblinProof(o.goblin_proof),
  };
}

export function toChonkVerify(o: MsgpackChonkVerify): ChonkVerify {
  if (o.proof === undefined) { throw new Error("Expected proof in ChonkVerify deserialization"); }
  if (o.vk === undefined) { throw new Error("Expected vk in ChonkVerify deserialization"); };
  return {
    proof: toChonkProof(o.proof),
    vk: o.vk,
  };
}

export function toVkAsFields(o: MsgpackVkAsFields): VkAsFields {
  if (o.verification_key === undefined) { throw new Error("Expected verification_key in VkAsFields deserialization"); };
  return {
    verificationKey: o.verification_key,
  };
}

export function toMegaVkAsFields(o: MsgpackMegaVkAsFields): MegaVkAsFields {
  if (o.verification_key === undefined) { throw new Error("Expected verification_key in MegaVkAsFields deserialization"); };
  return {
    verificationKey: o.verification_key,
  };
}

export function toCircuitWriteSolidityVerifier(o: MsgpackCircuitWriteSolidityVerifier): CircuitWriteSolidityVerifier {
  if (o.verification_key === undefined) { throw new Error("Expected verification_key in CircuitWriteSolidityVerifier deserialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitWriteSolidityVerifier deserialization"); };
  return {
    verificationKey: o.verification_key,
    settings: toProofSystemSettings(o.settings),
  };
}

export function toChonkCheckPrecomputedVk(o: MsgpackChonkCheckPrecomputedVk): ChonkCheckPrecomputedVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkCheckPrecomputedVk deserialization"); };
  return {
    circuit: toCircuitInput(o.circuit),
  };
}

export function toChonkStats(o: MsgpackChonkStats): ChonkStats {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkStats deserialization"); }
  if (o.include_gates_per_opcode === undefined) { throw new Error("Expected include_gates_per_opcode in ChonkStats deserialization"); };
  return {
    circuit: toCircuitInputNoVK(o.circuit),
    includeGatesPerOpcode: o.include_gates_per_opcode,
  };
}

export function toPoseidon2Hash(o: MsgpackPoseidon2Hash): Poseidon2Hash {
  if (o.inputs === undefined) { throw new Error("Expected inputs in Poseidon2Hash deserialization"); };
  return {
    inputs: o.inputs,
  };
}

export function toPoseidon2Permutation(o: MsgpackPoseidon2Permutation): Poseidon2Permutation {
  if (o.inputs === undefined) { throw new Error("Expected inputs in Poseidon2Permutation deserialization"); };
  return {
    inputs: o.inputs,
  };
}

export function toPoseidon2HashAccumulate(o: MsgpackPoseidon2HashAccumulate): Poseidon2HashAccumulate {
  if (o.inputs === undefined) { throw new Error("Expected inputs in Poseidon2HashAccumulate deserialization"); };
  return {
    inputs: o.inputs,
  };
}

export function toPedersenCommit(o: MsgpackPedersenCommit): PedersenCommit {
  if (o.inputs === undefined) { throw new Error("Expected inputs in PedersenCommit deserialization"); }
  if (o.hash_index === undefined) { throw new Error("Expected hash_index in PedersenCommit deserialization"); };
  return {
    inputs: o.inputs,
    hashIndex: o.hash_index,
  };
}

export function toPedersenHash(o: MsgpackPedersenHash): PedersenHash {
  if (o.inputs === undefined) { throw new Error("Expected inputs in PedersenHash deserialization"); }
  if (o.hash_index === undefined) { throw new Error("Expected hash_index in PedersenHash deserialization"); };
  return {
    inputs: o.inputs,
    hashIndex: o.hash_index,
  };
}

export function toPedersenHashBuffer(o: MsgpackPedersenHashBuffer): PedersenHashBuffer {
  if (o.input === undefined) { throw new Error("Expected input in PedersenHashBuffer deserialization"); }
  if (o.hash_index === undefined) { throw new Error("Expected hash_index in PedersenHashBuffer deserialization"); };
  return {
    input: o.input,
    hashIndex: o.hash_index,
  };
}

export function toBlake2s(o: MsgpackBlake2s): Blake2s {
  if (o.data === undefined) { throw new Error("Expected data in Blake2s deserialization"); };
  return {
    data: o.data,
  };
}

export function toBlake2sToField(o: MsgpackBlake2sToField): Blake2sToField {
  if (o.data === undefined) { throw new Error("Expected data in Blake2sToField deserialization"); };
  return {
    data: o.data,
  };
}

export function toAesEncrypt(o: MsgpackAesEncrypt): AesEncrypt {
  if (o.plaintext === undefined) { throw new Error("Expected plaintext in AesEncrypt deserialization"); }
  if (o.iv === undefined) { throw new Error("Expected iv in AesEncrypt deserialization"); }
  if (o.key === undefined) { throw new Error("Expected key in AesEncrypt deserialization"); }
  if (o.length === undefined) { throw new Error("Expected length in AesEncrypt deserialization"); };
  return {
    plaintext: o.plaintext,
    iv: o.iv,
    key: o.key,
    length: o.length,
  };
}

export function toAesDecrypt(o: MsgpackAesDecrypt): AesDecrypt {
  if (o.ciphertext === undefined) { throw new Error("Expected ciphertext in AesDecrypt deserialization"); }
  if (o.iv === undefined) { throw new Error("Expected iv in AesDecrypt deserialization"); }
  if (o.key === undefined) { throw new Error("Expected key in AesDecrypt deserialization"); }
  if (o.length === undefined) { throw new Error("Expected length in AesDecrypt deserialization"); };
  return {
    ciphertext: o.ciphertext,
    iv: o.iv,
    key: o.key,
    length: o.length,
  };
}

export function toGrumpkinPoint(o: MsgpackGrumpkinPoint): GrumpkinPoint {
  if (o.x === undefined) { throw new Error("Expected x in GrumpkinPoint deserialization"); }
  if (o.y === undefined) { throw new Error("Expected y in GrumpkinPoint deserialization"); };
  return {
    x: o.x,
    y: o.y,
  };
}

export function toGrumpkinMul(o: MsgpackGrumpkinMul): GrumpkinMul {
  if (o.point === undefined) { throw new Error("Expected point in GrumpkinMul deserialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in GrumpkinMul deserialization"); };
  return {
    point: toGrumpkinPoint(o.point),
    scalar: o.scalar,
  };
}

export function toGrumpkinAdd(o: MsgpackGrumpkinAdd): GrumpkinAdd {
  if (o.point_a === undefined) { throw new Error("Expected point_a in GrumpkinAdd deserialization"); }
  if (o.point_b === undefined) { throw new Error("Expected point_b in GrumpkinAdd deserialization"); };
  return {
    pointA: toGrumpkinPoint(o.point_a),
    pointB: toGrumpkinPoint(o.point_b),
  };
}

export function toGrumpkinBatchMul(o: MsgpackGrumpkinBatchMul): GrumpkinBatchMul {
  if (o.points === undefined) { throw new Error("Expected points in GrumpkinBatchMul deserialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in GrumpkinBatchMul deserialization"); };
  return {
    points: o.points.map((v: any) => toGrumpkinPoint(v)),
    scalar: o.scalar,
  };
}

export function toGrumpkinGetRandomFr(o: MsgpackGrumpkinGetRandomFr): GrumpkinGetRandomFr {
  if (o.dummy === undefined) { throw new Error("Expected dummy in GrumpkinGetRandomFr deserialization"); };
  return {
    dummy: o.dummy,
  };
}

export function toGrumpkinReduce512(o: MsgpackGrumpkinReduce512): GrumpkinReduce512 {
  if (o.input === undefined) { throw new Error("Expected input in GrumpkinReduce512 deserialization"); };
  return {
    input: o.input,
  };
}

export function toSecp256k1Point(o: MsgpackSecp256k1Point): Secp256k1Point {
  if (o.x === undefined) { throw new Error("Expected x in Secp256k1Point deserialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Secp256k1Point deserialization"); };
  return {
    x: o.x,
    y: o.y,
  };
}

export function toSecp256k1Mul(o: MsgpackSecp256k1Mul): Secp256k1Mul {
  if (o.point === undefined) { throw new Error("Expected point in Secp256k1Mul deserialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in Secp256k1Mul deserialization"); };
  return {
    point: toSecp256k1Point(o.point),
    scalar: o.scalar,
  };
}

export function toSecp256k1GetRandomFr(o: MsgpackSecp256k1GetRandomFr): Secp256k1GetRandomFr {
  if (o.dummy === undefined) { throw new Error("Expected dummy in Secp256k1GetRandomFr deserialization"); };
  return {
    dummy: o.dummy,
  };
}

export function toSecp256k1Reduce512(o: MsgpackSecp256k1Reduce512): Secp256k1Reduce512 {
  if (o.input === undefined) { throw new Error("Expected input in Secp256k1Reduce512 deserialization"); };
  return {
    input: o.input,
  };
}

export function toBn254FrSqrt(o: MsgpackBn254FrSqrt): Bn254FrSqrt {
  if (o.input === undefined) { throw new Error("Expected input in Bn254FrSqrt deserialization"); };
  return {
    input: o.input,
  };
}

export function toBn254FqSqrt(o: MsgpackBn254FqSqrt): Bn254FqSqrt {
  if (o.input === undefined) { throw new Error("Expected input in Bn254FqSqrt deserialization"); };
  return {
    input: o.input,
  };
}

export function toBn254G1Point(o: MsgpackBn254G1Point): Bn254G1Point {
  if (o.x === undefined) { throw new Error("Expected x in Bn254G1Point deserialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Bn254G1Point deserialization"); };
  return {
    x: o.x,
    y: o.y,
  };
}

export function toBn254G1Mul(o: MsgpackBn254G1Mul): Bn254G1Mul {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1Mul deserialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in Bn254G1Mul deserialization"); };
  return {
    point: toBn254G1Point(o.point),
    scalar: o.scalar,
  };
}

export function toBn254G2Point(o: MsgpackBn254G2Point): Bn254G2Point {
  if (o.x === undefined) { throw new Error("Expected x in Bn254G2Point deserialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Bn254G2Point deserialization"); };
  return {
    x: o.x,
    y: o.y,
  };
}

export function toBn254G2Mul(o: MsgpackBn254G2Mul): Bn254G2Mul {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G2Mul deserialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in Bn254G2Mul deserialization"); };
  return {
    point: toBn254G2Point(o.point),
    scalar: o.scalar,
  };
}

export function toBn254G1IsOnCurve(o: MsgpackBn254G1IsOnCurve): Bn254G1IsOnCurve {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1IsOnCurve deserialization"); };
  return {
    point: toBn254G1Point(o.point),
  };
}

export function toBn254G1FromCompressed(o: MsgpackBn254G1FromCompressed): Bn254G1FromCompressed {
  if (o.compressed === undefined) { throw new Error("Expected compressed in Bn254G1FromCompressed deserialization"); };
  return {
    compressed: o.compressed,
  };
}

export function toSchnorrComputePublicKey(o: MsgpackSchnorrComputePublicKey): SchnorrComputePublicKey {
  if (o.private_key === undefined) { throw new Error("Expected private_key in SchnorrComputePublicKey deserialization"); };
  return {
    privateKey: o.private_key,
  };
}

export function toSchnorrConstructSignature(o: MsgpackSchnorrConstructSignature): SchnorrConstructSignature {
  if (o.message === undefined) { throw new Error("Expected message in SchnorrConstructSignature deserialization"); }
  if (o.private_key === undefined) { throw new Error("Expected private_key in SchnorrConstructSignature deserialization"); };
  return {
    message: o.message,
    privateKey: o.private_key,
  };
}

export function toSchnorrVerifySignature(o: MsgpackSchnorrVerifySignature): SchnorrVerifySignature {
  if (o.message === undefined) { throw new Error("Expected message in SchnorrVerifySignature deserialization"); }
  if (o.public_key === undefined) { throw new Error("Expected public_key in SchnorrVerifySignature deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in SchnorrVerifySignature deserialization"); }
  if (o.e === undefined) { throw new Error("Expected e in SchnorrVerifySignature deserialization"); };
  return {
    message: o.message,
    publicKey: toGrumpkinPoint(o.public_key),
    s: o.s,
    e: o.e,
  };
}

export function toEcdsaSecp256k1ComputePublicKey(o: MsgpackEcdsaSecp256k1ComputePublicKey): EcdsaSecp256k1ComputePublicKey {
  if (o.private_key === undefined) { throw new Error("Expected private_key in EcdsaSecp256k1ComputePublicKey deserialization"); };
  return {
    privateKey: o.private_key,
  };
}

export function toEcdsaSecp256r1ComputePublicKey(o: MsgpackEcdsaSecp256r1ComputePublicKey): EcdsaSecp256r1ComputePublicKey {
  if (o.private_key === undefined) { throw new Error("Expected private_key in EcdsaSecp256r1ComputePublicKey deserialization"); };
  return {
    privateKey: o.private_key,
  };
}

export function toEcdsaSecp256k1ConstructSignature(o: MsgpackEcdsaSecp256k1ConstructSignature): EcdsaSecp256k1ConstructSignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256k1ConstructSignature deserialization"); }
  if (o.private_key === undefined) { throw new Error("Expected private_key in EcdsaSecp256k1ConstructSignature deserialization"); };
  return {
    message: o.message,
    privateKey: o.private_key,
  };
}

export function toEcdsaSecp256r1ConstructSignature(o: MsgpackEcdsaSecp256r1ConstructSignature): EcdsaSecp256r1ConstructSignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256r1ConstructSignature deserialization"); }
  if (o.private_key === undefined) { throw new Error("Expected private_key in EcdsaSecp256r1ConstructSignature deserialization"); };
  return {
    message: o.message,
    privateKey: o.private_key,
  };
}

export function toEcdsaSecp256k1RecoverPublicKey(o: MsgpackEcdsaSecp256k1RecoverPublicKey): EcdsaSecp256k1RecoverPublicKey {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256k1RecoverPublicKey deserialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256k1RecoverPublicKey deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256k1RecoverPublicKey deserialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256k1RecoverPublicKey deserialization"); };
  return {
    message: o.message,
    r: o.r,
    s: o.s,
    v: o.v,
  };
}

export function toEcdsaSecp256r1RecoverPublicKey(o: MsgpackEcdsaSecp256r1RecoverPublicKey): EcdsaSecp256r1RecoverPublicKey {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256r1RecoverPublicKey deserialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256r1RecoverPublicKey deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256r1RecoverPublicKey deserialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256r1RecoverPublicKey deserialization"); };
  return {
    message: o.message,
    r: o.r,
    s: o.s,
    v: o.v,
  };
}

export function toEcdsaSecp256k1VerifySignature(o: MsgpackEcdsaSecp256k1VerifySignature): EcdsaSecp256k1VerifySignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256k1VerifySignature deserialization"); }
  if (o.public_key === undefined) { throw new Error("Expected public_key in EcdsaSecp256k1VerifySignature deserialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256k1VerifySignature deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256k1VerifySignature deserialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256k1VerifySignature deserialization"); };
  return {
    message: o.message,
    publicKey: toSecp256k1Point(o.public_key),
    r: o.r,
    s: o.s,
    v: o.v,
  };
}

export function toSecp256r1Point(o: MsgpackSecp256r1Point): Secp256r1Point {
  if (o.x === undefined) { throw new Error("Expected x in Secp256r1Point deserialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Secp256r1Point deserialization"); };
  return {
    x: o.x,
    y: o.y,
  };
}

export function toEcdsaSecp256r1VerifySignature(o: MsgpackEcdsaSecp256r1VerifySignature): EcdsaSecp256r1VerifySignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256r1VerifySignature deserialization"); }
  if (o.public_key === undefined) { throw new Error("Expected public_key in EcdsaSecp256r1VerifySignature deserialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256r1VerifySignature deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256r1VerifySignature deserialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256r1VerifySignature deserialization"); };
  return {
    message: o.message,
    publicKey: toSecp256r1Point(o.public_key),
    r: o.r,
    s: o.s,
    v: o.v,
  };
}

export function toSrsInitSrs(o: MsgpackSrsInitSrs): SrsInitSrs {
  if (o.points_buf === undefined) { throw new Error("Expected points_buf in SrsInitSrs deserialization"); }
  if (o.num_points === undefined) { throw new Error("Expected num_points in SrsInitSrs deserialization"); }
  if (o.g2_point === undefined) { throw new Error("Expected g2_point in SrsInitSrs deserialization"); };
  return {
    pointsBuf: o.points_buf,
    numPoints: o.num_points,
    g2Point: o.g2_point,
  };
}

export function toSrsInitGrumpkinSrs(o: MsgpackSrsInitGrumpkinSrs): SrsInitGrumpkinSrs {
  if (o.points_buf === undefined) { throw new Error("Expected points_buf in SrsInitGrumpkinSrs deserialization"); }
  if (o.num_points === undefined) { throw new Error("Expected num_points in SrsInitGrumpkinSrs deserialization"); };
  return {
    pointsBuf: o.points_buf,
    numPoints: o.num_points,
  };
}

export function toShutdown(o: MsgpackShutdown): Shutdown {
  return {};
}

export function toErrorResponse(o: MsgpackErrorResponse): ErrorResponse {
  if (o.message === undefined) { throw new Error("Expected message in ErrorResponse deserialization"); };
  return {
    message: o.message,
  };
}

export function toCircuitProveResponse(o: MsgpackCircuitProveResponse): CircuitProveResponse {
  if (o.public_inputs === undefined) { throw new Error("Expected public_inputs in CircuitProveResponse deserialization"); }
  if (o.proof === undefined) { throw new Error("Expected proof in CircuitProveResponse deserialization"); }
  if (o.vk === undefined) { throw new Error("Expected vk in CircuitProveResponse deserialization"); };
  return {
    publicInputs: o.public_inputs,
    proof: o.proof,
    vk: toCircuitComputeVkResponse(o.vk),
  };
}

export function toCircuitInfoResponse(o: MsgpackCircuitInfoResponse): CircuitInfoResponse {
  if (o.num_gates === undefined) { throw new Error("Expected num_gates in CircuitInfoResponse deserialization"); }
  if (o.num_gates_dyadic === undefined) { throw new Error("Expected num_gates_dyadic in CircuitInfoResponse deserialization"); }
  if (o.num_acir_opcodes === undefined) { throw new Error("Expected num_acir_opcodes in CircuitInfoResponse deserialization"); }
  if (o.gates_per_opcode === undefined) { throw new Error("Expected gates_per_opcode in CircuitInfoResponse deserialization"); };
  return {
    numGates: o.num_gates,
    numGatesDyadic: o.num_gates_dyadic,
    numAcirOpcodes: o.num_acir_opcodes,
    gatesPerOpcode: o.gates_per_opcode,
  };
}

export function toCircuitVerifyResponse(o: MsgpackCircuitVerifyResponse): CircuitVerifyResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in CircuitVerifyResponse deserialization"); };
  return {
    verified: o.verified,
  };
}

export function toChonkComputeStandaloneVkResponse(o: MsgpackChonkComputeStandaloneVkResponse): ChonkComputeStandaloneVkResponse {
  if (o.bytes === undefined) { throw new Error("Expected bytes in ChonkComputeStandaloneVkResponse deserialization"); }
  if (o.fields === undefined) { throw new Error("Expected fields in ChonkComputeStandaloneVkResponse deserialization"); };
  return {
    bytes: o.bytes,
    fields: o.fields,
  };
}

export function toChonkComputeIvcVkResponse(o: MsgpackChonkComputeIvcVkResponse): ChonkComputeIvcVkResponse {
  if (o.bytes === undefined) { throw new Error("Expected bytes in ChonkComputeIvcVkResponse deserialization"); };
  return {
    bytes: o.bytes,
  };
}

export function toChonkStartResponse(o: MsgpackChonkStartResponse): ChonkStartResponse {
  return {};
}

export function toChonkLoadResponse(o: MsgpackChonkLoadResponse): ChonkLoadResponse {
  return {};
}

export function toChonkAccumulateResponse(o: MsgpackChonkAccumulateResponse): ChonkAccumulateResponse {
  return {};
}

export function toChonkProveResponse(o: MsgpackChonkProveResponse): ChonkProveResponse {
  if (o.proof === undefined) { throw new Error("Expected proof in ChonkProveResponse deserialization"); };
  return {
    proof: toChonkProof(o.proof),
  };
}

export function toChonkVerifyResponse(o: MsgpackChonkVerifyResponse): ChonkVerifyResponse {
  if (o.valid === undefined) { throw new Error("Expected valid in ChonkVerifyResponse deserialization"); };
  return {
    valid: o.valid,
  };
}

export function toVkAsFieldsResponse(o: MsgpackVkAsFieldsResponse): VkAsFieldsResponse {
  if (o.fields === undefined) { throw new Error("Expected fields in VkAsFieldsResponse deserialization"); };
  return {
    fields: o.fields,
  };
}

export function toMegaVkAsFieldsResponse(o: MsgpackMegaVkAsFieldsResponse): MegaVkAsFieldsResponse {
  if (o.fields === undefined) { throw new Error("Expected fields in MegaVkAsFieldsResponse deserialization"); };
  return {
    fields: o.fields,
  };
}

export function toCircuitWriteSolidityVerifierResponse(o: MsgpackCircuitWriteSolidityVerifierResponse): CircuitWriteSolidityVerifierResponse {
  if (o.solidity_code === undefined) { throw new Error("Expected solidity_code in CircuitWriteSolidityVerifierResponse deserialization"); };
  return {
    solidityCode: o.solidity_code,
  };
}

export function toChonkCheckPrecomputedVkResponse(o: MsgpackChonkCheckPrecomputedVkResponse): ChonkCheckPrecomputedVkResponse {
  if (o.valid === undefined) { throw new Error("Expected valid in ChonkCheckPrecomputedVkResponse deserialization"); }
  if (o.actual_vk === undefined) { throw new Error("Expected actual_vk in ChonkCheckPrecomputedVkResponse deserialization"); };
  return {
    valid: o.valid,
    actualVk: o.actual_vk,
  };
}

export function toChonkStatsResponse(o: MsgpackChonkStatsResponse): ChonkStatsResponse {
  if (o.acir_opcodes === undefined) { throw new Error("Expected acir_opcodes in ChonkStatsResponse deserialization"); }
  if (o.circuit_size === undefined) { throw new Error("Expected circuit_size in ChonkStatsResponse deserialization"); }
  if (o.gates_per_opcode === undefined) { throw new Error("Expected gates_per_opcode in ChonkStatsResponse deserialization"); };
  return {
    acirOpcodes: o.acir_opcodes,
    circuitSize: o.circuit_size,
    gatesPerOpcode: o.gates_per_opcode,
  };
}

export function toPoseidon2HashResponse(o: MsgpackPoseidon2HashResponse): Poseidon2HashResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in Poseidon2HashResponse deserialization"); };
  return {
    hash: o.hash,
  };
}

export function toPoseidon2PermutationResponse(o: MsgpackPoseidon2PermutationResponse): Poseidon2PermutationResponse {
  if (o.outputs === undefined) { throw new Error("Expected outputs in Poseidon2PermutationResponse deserialization"); };
  return {
    outputs: o.outputs,
  };
}

export function toPoseidon2HashAccumulateResponse(o: MsgpackPoseidon2HashAccumulateResponse): Poseidon2HashAccumulateResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in Poseidon2HashAccumulateResponse deserialization"); };
  return {
    hash: o.hash,
  };
}

export function toPedersenCommitResponse(o: MsgpackPedersenCommitResponse): PedersenCommitResponse {
  if (o.point === undefined) { throw new Error("Expected point in PedersenCommitResponse deserialization"); };
  return {
    point: toGrumpkinPoint(o.point),
  };
}

export function toPedersenHashResponse(o: MsgpackPedersenHashResponse): PedersenHashResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in PedersenHashResponse deserialization"); };
  return {
    hash: o.hash,
  };
}

export function toPedersenHashBufferResponse(o: MsgpackPedersenHashBufferResponse): PedersenHashBufferResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in PedersenHashBufferResponse deserialization"); };
  return {
    hash: o.hash,
  };
}

export function toBlake2sResponse(o: MsgpackBlake2sResponse): Blake2sResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in Blake2sResponse deserialization"); };
  return {
    hash: o.hash,
  };
}

export function toBlake2sToFieldResponse(o: MsgpackBlake2sToFieldResponse): Blake2sToFieldResponse {
  if (o.field === undefined) { throw new Error("Expected field in Blake2sToFieldResponse deserialization"); };
  return {
    field: o.field,
  };
}

export function toAesEncryptResponse(o: MsgpackAesEncryptResponse): AesEncryptResponse {
  if (o.ciphertext === undefined) { throw new Error("Expected ciphertext in AesEncryptResponse deserialization"); };
  return {
    ciphertext: o.ciphertext,
  };
}

export function toAesDecryptResponse(o: MsgpackAesDecryptResponse): AesDecryptResponse {
  if (o.plaintext === undefined) { throw new Error("Expected plaintext in AesDecryptResponse deserialization"); };
  return {
    plaintext: o.plaintext,
  };
}

export function toGrumpkinMulResponse(o: MsgpackGrumpkinMulResponse): GrumpkinMulResponse {
  if (o.point === undefined) { throw new Error("Expected point in GrumpkinMulResponse deserialization"); };
  return {
    point: toGrumpkinPoint(o.point),
  };
}

export function toGrumpkinAddResponse(o: MsgpackGrumpkinAddResponse): GrumpkinAddResponse {
  if (o.point === undefined) { throw new Error("Expected point in GrumpkinAddResponse deserialization"); };
  return {
    point: toGrumpkinPoint(o.point),
  };
}

export function toGrumpkinBatchMulResponse(o: MsgpackGrumpkinBatchMulResponse): GrumpkinBatchMulResponse {
  if (o.points === undefined) { throw new Error("Expected points in GrumpkinBatchMulResponse deserialization"); };
  return {
    points: o.points.map((v: any) => toGrumpkinPoint(v)),
  };
}

export function toGrumpkinGetRandomFrResponse(o: MsgpackGrumpkinGetRandomFrResponse): GrumpkinGetRandomFrResponse {
  if (o.value === undefined) { throw new Error("Expected value in GrumpkinGetRandomFrResponse deserialization"); };
  return {
    value: o.value,
  };
}

export function toGrumpkinReduce512Response(o: MsgpackGrumpkinReduce512Response): GrumpkinReduce512Response {
  if (o.value === undefined) { throw new Error("Expected value in GrumpkinReduce512Response deserialization"); };
  return {
    value: o.value,
  };
}

export function toSecp256k1MulResponse(o: MsgpackSecp256k1MulResponse): Secp256k1MulResponse {
  if (o.point === undefined) { throw new Error("Expected point in Secp256k1MulResponse deserialization"); };
  return {
    point: toSecp256k1Point(o.point),
  };
}

export function toSecp256k1GetRandomFrResponse(o: MsgpackSecp256k1GetRandomFrResponse): Secp256k1GetRandomFrResponse {
  if (o.value === undefined) { throw new Error("Expected value in Secp256k1GetRandomFrResponse deserialization"); };
  return {
    value: o.value,
  };
}

export function toSecp256k1Reduce512Response(o: MsgpackSecp256k1Reduce512Response): Secp256k1Reduce512Response {
  if (o.value === undefined) { throw new Error("Expected value in Secp256k1Reduce512Response deserialization"); };
  return {
    value: o.value,
  };
}

export function toBn254FrSqrtResponse(o: MsgpackBn254FrSqrtResponse): Bn254FrSqrtResponse {
  if (o.is_square_root === undefined) { throw new Error("Expected is_square_root in Bn254FrSqrtResponse deserialization"); }
  if (o.value === undefined) { throw new Error("Expected value in Bn254FrSqrtResponse deserialization"); };
  return {
    isSquareRoot: o.is_square_root,
    value: o.value,
  };
}

export function toBn254FqSqrtResponse(o: MsgpackBn254FqSqrtResponse): Bn254FqSqrtResponse {
  if (o.is_square_root === undefined) { throw new Error("Expected is_square_root in Bn254FqSqrtResponse deserialization"); }
  if (o.value === undefined) { throw new Error("Expected value in Bn254FqSqrtResponse deserialization"); };
  return {
    isSquareRoot: o.is_square_root,
    value: o.value,
  };
}

export function toBn254G1MulResponse(o: MsgpackBn254G1MulResponse): Bn254G1MulResponse {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1MulResponse deserialization"); };
  return {
    point: toBn254G1Point(o.point),
  };
}

export function toBn254G2MulResponse(o: MsgpackBn254G2MulResponse): Bn254G2MulResponse {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G2MulResponse deserialization"); };
  return {
    point: toBn254G2Point(o.point),
  };
}

export function toBn254G1IsOnCurveResponse(o: MsgpackBn254G1IsOnCurveResponse): Bn254G1IsOnCurveResponse {
  if (o.is_on_curve === undefined) { throw new Error("Expected is_on_curve in Bn254G1IsOnCurveResponse deserialization"); };
  return {
    isOnCurve: o.is_on_curve,
  };
}

export function toBn254G1FromCompressedResponse(o: MsgpackBn254G1FromCompressedResponse): Bn254G1FromCompressedResponse {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1FromCompressedResponse deserialization"); };
  return {
    point: toBn254G1Point(o.point),
  };
}

export function toSchnorrComputePublicKeyResponse(o: MsgpackSchnorrComputePublicKeyResponse): SchnorrComputePublicKeyResponse {
  if (o.public_key === undefined) { throw new Error("Expected public_key in SchnorrComputePublicKeyResponse deserialization"); };
  return {
    publicKey: toGrumpkinPoint(o.public_key),
  };
}

export function toSchnorrConstructSignatureResponse(o: MsgpackSchnorrConstructSignatureResponse): SchnorrConstructSignatureResponse {
  if (o.s === undefined) { throw new Error("Expected s in SchnorrConstructSignatureResponse deserialization"); }
  if (o.e === undefined) { throw new Error("Expected e in SchnorrConstructSignatureResponse deserialization"); };
  return {
    s: o.s,
    e: o.e,
  };
}

export function toSchnorrVerifySignatureResponse(o: MsgpackSchnorrVerifySignatureResponse): SchnorrVerifySignatureResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in SchnorrVerifySignatureResponse deserialization"); };
  return {
    verified: o.verified,
  };
}

export function toEcdsaSecp256k1ComputePublicKeyResponse(o: MsgpackEcdsaSecp256k1ComputePublicKeyResponse): EcdsaSecp256k1ComputePublicKeyResponse {
  if (o.public_key === undefined) { throw new Error("Expected public_key in EcdsaSecp256k1ComputePublicKeyResponse deserialization"); };
  return {
    publicKey: toSecp256k1Point(o.public_key),
  };
}

export function toEcdsaSecp256r1ComputePublicKeyResponse(o: MsgpackEcdsaSecp256r1ComputePublicKeyResponse): EcdsaSecp256r1ComputePublicKeyResponse {
  if (o.public_key === undefined) { throw new Error("Expected public_key in EcdsaSecp256r1ComputePublicKeyResponse deserialization"); };
  return {
    publicKey: toSecp256r1Point(o.public_key),
  };
}

export function toEcdsaSecp256k1ConstructSignatureResponse(o: MsgpackEcdsaSecp256k1ConstructSignatureResponse): EcdsaSecp256k1ConstructSignatureResponse {
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256k1ConstructSignatureResponse deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256k1ConstructSignatureResponse deserialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256k1ConstructSignatureResponse deserialization"); };
  return {
    r: o.r,
    s: o.s,
    v: o.v,
  };
}

export function toEcdsaSecp256r1ConstructSignatureResponse(o: MsgpackEcdsaSecp256r1ConstructSignatureResponse): EcdsaSecp256r1ConstructSignatureResponse {
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256r1ConstructSignatureResponse deserialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256r1ConstructSignatureResponse deserialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256r1ConstructSignatureResponse deserialization"); };
  return {
    r: o.r,
    s: o.s,
    v: o.v,
  };
}

export function toEcdsaSecp256k1RecoverPublicKeyResponse(o: MsgpackEcdsaSecp256k1RecoverPublicKeyResponse): EcdsaSecp256k1RecoverPublicKeyResponse {
  if (o.public_key === undefined) { throw new Error("Expected public_key in EcdsaSecp256k1RecoverPublicKeyResponse deserialization"); };
  return {
    publicKey: toSecp256k1Point(o.public_key),
  };
}

export function toEcdsaSecp256r1RecoverPublicKeyResponse(o: MsgpackEcdsaSecp256r1RecoverPublicKeyResponse): EcdsaSecp256r1RecoverPublicKeyResponse {
  if (o.public_key === undefined) { throw new Error("Expected public_key in EcdsaSecp256r1RecoverPublicKeyResponse deserialization"); };
  return {
    publicKey: toSecp256r1Point(o.public_key),
  };
}

export function toEcdsaSecp256k1VerifySignatureResponse(o: MsgpackEcdsaSecp256k1VerifySignatureResponse): EcdsaSecp256k1VerifySignatureResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in EcdsaSecp256k1VerifySignatureResponse deserialization"); };
  return {
    verified: o.verified,
  };
}

export function toEcdsaSecp256r1VerifySignatureResponse(o: MsgpackEcdsaSecp256r1VerifySignatureResponse): EcdsaSecp256r1VerifySignatureResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in EcdsaSecp256r1VerifySignatureResponse deserialization"); };
  return {
    verified: o.verified,
  };
}

export function toSrsInitSrsResponse(o: MsgpackSrsInitSrsResponse): SrsInitSrsResponse {
  if (o.dummy === undefined) { throw new Error("Expected dummy in SrsInitSrsResponse deserialization"); };
  return {
    dummy: o.dummy,
  };
}

export function toSrsInitGrumpkinSrsResponse(o: MsgpackSrsInitGrumpkinSrsResponse): SrsInitGrumpkinSrsResponse {
  if (o.dummy === undefined) { throw new Error("Expected dummy in SrsInitGrumpkinSrsResponse deserialization"); };
  return {
    dummy: o.dummy,
  };
}

export function toShutdownResponse(o: MsgpackShutdownResponse): ShutdownResponse {
  return {};
}

export function fromCircuitComputeVkResponse(o: CircuitComputeVkResponse): MsgpackCircuitComputeVkResponse {
  if (o.bytes === undefined) { throw new Error("Expected bytes in CircuitComputeVkResponse serialization"); }
  if (o.fields === undefined) { throw new Error("Expected fields in CircuitComputeVkResponse serialization"); }
  if (o.hash === undefined) { throw new Error("Expected hash in CircuitComputeVkResponse serialization"); };
  return {
  bytes: o.bytes,
  fields: o.fields,
  hash: o.hash,
  };
}

export function fromCircuitInput(o: CircuitInput): MsgpackCircuitInput {
  if (o.name === undefined) { throw new Error("Expected name in CircuitInput serialization"); }
  if (o.bytecode === undefined) { throw new Error("Expected bytecode in CircuitInput serialization"); }
  if (o.verificationKey === undefined) { throw new Error("Expected verificationKey in CircuitInput serialization"); };
  return {
  name: o.name,
  bytecode: o.bytecode,
  verification_key: o.verificationKey,
  };
}

export function fromProofSystemSettings(o: ProofSystemSettings): MsgpackProofSystemSettings {
  if (o.ipaAccumulation === undefined) { throw new Error("Expected ipaAccumulation in ProofSystemSettings serialization"); }
  if (o.oracleHashType === undefined) { throw new Error("Expected oracleHashType in ProofSystemSettings serialization"); }
  if (o.disableZk === undefined) { throw new Error("Expected disableZk in ProofSystemSettings serialization"); }
  if (o.optimizedSolidityVerifier === undefined) { throw new Error("Expected optimizedSolidityVerifier in ProofSystemSettings serialization"); };
  return {
  ipa_accumulation: o.ipaAccumulation,
  oracle_hash_type: o.oracleHashType,
  disable_zk: o.disableZk,
  optimized_solidity_verifier: o.optimizedSolidityVerifier,
  };
}

export function fromCircuitProve(o: CircuitProve): MsgpackCircuitProve {
  if (o.circuit === undefined) { throw new Error("Expected circuit in CircuitProve serialization"); }
  if (o.witness === undefined) { throw new Error("Expected witness in CircuitProve serialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitProve serialization"); };
  return {
  circuit: fromCircuitInput(o.circuit),
  witness: o.witness,
  settings: fromProofSystemSettings(o.settings),
  };
}

export function fromCircuitInputNoVK(o: CircuitInputNoVK): MsgpackCircuitInputNoVK {
  if (o.name === undefined) { throw new Error("Expected name in CircuitInputNoVK serialization"); }
  if (o.bytecode === undefined) { throw new Error("Expected bytecode in CircuitInputNoVK serialization"); };
  return {
  name: o.name,
  bytecode: o.bytecode,
  };
}

export function fromCircuitComputeVk(o: CircuitComputeVk): MsgpackCircuitComputeVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in CircuitComputeVk serialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitComputeVk serialization"); };
  return {
  circuit: fromCircuitInputNoVK(o.circuit),
  settings: fromProofSystemSettings(o.settings),
  };
}

export function fromCircuitStats(o: CircuitStats): MsgpackCircuitStats {
  if (o.circuit === undefined) { throw new Error("Expected circuit in CircuitStats serialization"); }
  if (o.includeGatesPerOpcode === undefined) { throw new Error("Expected includeGatesPerOpcode in CircuitStats serialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitStats serialization"); };
  return {
  circuit: fromCircuitInput(o.circuit),
  include_gates_per_opcode: o.includeGatesPerOpcode,
  settings: fromProofSystemSettings(o.settings),
  };
}

export function fromCircuitVerify(o: CircuitVerify): MsgpackCircuitVerify {
  if (o.verificationKey === undefined) { throw new Error("Expected verificationKey in CircuitVerify serialization"); }
  if (o.publicInputs === undefined) { throw new Error("Expected publicInputs in CircuitVerify serialization"); }
  if (o.proof === undefined) { throw new Error("Expected proof in CircuitVerify serialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitVerify serialization"); };
  return {
  verification_key: o.verificationKey,
  public_inputs: o.publicInputs,
  proof: o.proof,
  settings: fromProofSystemSettings(o.settings),
  };
}

export function fromChonkComputeStandaloneVk(o: ChonkComputeStandaloneVk): MsgpackChonkComputeStandaloneVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkComputeStandaloneVk serialization"); };
  return {
  circuit: fromCircuitInputNoVK(o.circuit),
  };
}

export function fromChonkComputeIvcVk(o: ChonkComputeIvcVk): MsgpackChonkComputeIvcVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkComputeIvcVk serialization"); };
  return {
  circuit: fromCircuitInputNoVK(o.circuit),
  };
}

export function fromChonkStart(o: ChonkStart): MsgpackChonkStart {
  if (o.numCircuits === undefined) { throw new Error("Expected numCircuits in ChonkStart serialization"); };
  return {
  num_circuits: o.numCircuits,
  };
}

export function fromChonkLoad(o: ChonkLoad): MsgpackChonkLoad {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkLoad serialization"); };
  return {
  circuit: fromCircuitInput(o.circuit),
  };
}

export function fromChonkAccumulate(o: ChonkAccumulate): MsgpackChonkAccumulate {
  if (o.witness === undefined) { throw new Error("Expected witness in ChonkAccumulate serialization"); };
  return {
  witness: o.witness,
  };
}

export function fromChonkProve(o: ChonkProve): MsgpackChonkProve {
  return {};
}

export function fromGoblinProof(o: GoblinProof): MsgpackGoblinProof {
  if (o.mergeProof === undefined) { throw new Error("Expected mergeProof in GoblinProof serialization"); }
  if (o.eccvmProof === undefined) { throw new Error("Expected eccvmProof in GoblinProof serialization"); }
  if (o.ipaProof === undefined) { throw new Error("Expected ipaProof in GoblinProof serialization"); }
  if (o.translatorProof === undefined) { throw new Error("Expected translatorProof in GoblinProof serialization"); };
  return {
  merge_proof: o.mergeProof,
  eccvm_proof: o.eccvmProof,
  ipa_proof: o.ipaProof,
  translator_proof: o.translatorProof,
  };
}

export function fromChonkProof(o: ChonkProof): MsgpackChonkProof {
  if (o.megaProof === undefined) { throw new Error("Expected megaProof in ChonkProof serialization"); }
  if (o.goblinProof === undefined) { throw new Error("Expected goblinProof in ChonkProof serialization"); };
  return {
  mega_proof: o.megaProof,
  goblin_proof: fromGoblinProof(o.goblinProof),
  };
}

export function fromChonkVerify(o: ChonkVerify): MsgpackChonkVerify {
  if (o.proof === undefined) { throw new Error("Expected proof in ChonkVerify serialization"); }
  if (o.vk === undefined) { throw new Error("Expected vk in ChonkVerify serialization"); };
  return {
  proof: fromChonkProof(o.proof),
  vk: o.vk,
  };
}

export function fromVkAsFields(o: VkAsFields): MsgpackVkAsFields {
  if (o.verificationKey === undefined) { throw new Error("Expected verificationKey in VkAsFields serialization"); };
  return {
  verification_key: o.verificationKey,
  };
}

export function fromMegaVkAsFields(o: MegaVkAsFields): MsgpackMegaVkAsFields {
  if (o.verificationKey === undefined) { throw new Error("Expected verificationKey in MegaVkAsFields serialization"); };
  return {
  verification_key: o.verificationKey,
  };
}

export function fromCircuitWriteSolidityVerifier(o: CircuitWriteSolidityVerifier): MsgpackCircuitWriteSolidityVerifier {
  if (o.verificationKey === undefined) { throw new Error("Expected verificationKey in CircuitWriteSolidityVerifier serialization"); }
  if (o.settings === undefined) { throw new Error("Expected settings in CircuitWriteSolidityVerifier serialization"); };
  return {
  verification_key: o.verificationKey,
  settings: fromProofSystemSettings(o.settings),
  };
}

export function fromChonkCheckPrecomputedVk(o: ChonkCheckPrecomputedVk): MsgpackChonkCheckPrecomputedVk {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkCheckPrecomputedVk serialization"); };
  return {
  circuit: fromCircuitInput(o.circuit),
  };
}

export function fromChonkStats(o: ChonkStats): MsgpackChonkStats {
  if (o.circuit === undefined) { throw new Error("Expected circuit in ChonkStats serialization"); }
  if (o.includeGatesPerOpcode === undefined) { throw new Error("Expected includeGatesPerOpcode in ChonkStats serialization"); };
  return {
  circuit: fromCircuitInputNoVK(o.circuit),
  include_gates_per_opcode: o.includeGatesPerOpcode,
  };
}

export function fromPoseidon2Hash(o: Poseidon2Hash): MsgpackPoseidon2Hash {
  if (o.inputs === undefined) { throw new Error("Expected inputs in Poseidon2Hash serialization"); };
  return {
  inputs: o.inputs,
  };
}

export function fromPoseidon2Permutation(o: Poseidon2Permutation): MsgpackPoseidon2Permutation {
  if (o.inputs === undefined) { throw new Error("Expected inputs in Poseidon2Permutation serialization"); };
  return {
  inputs: o.inputs,
  };
}

export function fromPoseidon2HashAccumulate(o: Poseidon2HashAccumulate): MsgpackPoseidon2HashAccumulate {
  if (o.inputs === undefined) { throw new Error("Expected inputs in Poseidon2HashAccumulate serialization"); };
  return {
  inputs: o.inputs,
  };
}

export function fromPedersenCommit(o: PedersenCommit): MsgpackPedersenCommit {
  if (o.inputs === undefined) { throw new Error("Expected inputs in PedersenCommit serialization"); }
  if (o.hashIndex === undefined) { throw new Error("Expected hashIndex in PedersenCommit serialization"); };
  return {
  inputs: o.inputs,
  hash_index: o.hashIndex,
  };
}

export function fromPedersenHash(o: PedersenHash): MsgpackPedersenHash {
  if (o.inputs === undefined) { throw new Error("Expected inputs in PedersenHash serialization"); }
  if (o.hashIndex === undefined) { throw new Error("Expected hashIndex in PedersenHash serialization"); };
  return {
  inputs: o.inputs,
  hash_index: o.hashIndex,
  };
}

export function fromPedersenHashBuffer(o: PedersenHashBuffer): MsgpackPedersenHashBuffer {
  if (o.input === undefined) { throw new Error("Expected input in PedersenHashBuffer serialization"); }
  if (o.hashIndex === undefined) { throw new Error("Expected hashIndex in PedersenHashBuffer serialization"); };
  return {
  input: o.input,
  hash_index: o.hashIndex,
  };
}

export function fromBlake2s(o: Blake2s): MsgpackBlake2s {
  if (o.data === undefined) { throw new Error("Expected data in Blake2s serialization"); };
  return {
  data: o.data,
  };
}

export function fromBlake2sToField(o: Blake2sToField): MsgpackBlake2sToField {
  if (o.data === undefined) { throw new Error("Expected data in Blake2sToField serialization"); };
  return {
  data: o.data,
  };
}

export function fromAesEncrypt(o: AesEncrypt): MsgpackAesEncrypt {
  if (o.plaintext === undefined) { throw new Error("Expected plaintext in AesEncrypt serialization"); }
  if (o.iv === undefined) { throw new Error("Expected iv in AesEncrypt serialization"); }
  if (o.key === undefined) { throw new Error("Expected key in AesEncrypt serialization"); }
  if (o.length === undefined) { throw new Error("Expected length in AesEncrypt serialization"); };
  return {
  plaintext: o.plaintext,
  iv: o.iv,
  key: o.key,
  length: o.length,
  };
}

export function fromAesDecrypt(o: AesDecrypt): MsgpackAesDecrypt {
  if (o.ciphertext === undefined) { throw new Error("Expected ciphertext in AesDecrypt serialization"); }
  if (o.iv === undefined) { throw new Error("Expected iv in AesDecrypt serialization"); }
  if (o.key === undefined) { throw new Error("Expected key in AesDecrypt serialization"); }
  if (o.length === undefined) { throw new Error("Expected length in AesDecrypt serialization"); };
  return {
  ciphertext: o.ciphertext,
  iv: o.iv,
  key: o.key,
  length: o.length,
  };
}

export function fromGrumpkinPoint(o: GrumpkinPoint): MsgpackGrumpkinPoint {
  if (o.x === undefined) { throw new Error("Expected x in GrumpkinPoint serialization"); }
  if (o.y === undefined) { throw new Error("Expected y in GrumpkinPoint serialization"); };
  return {
  x: o.x,
  y: o.y,
  };
}

export function fromGrumpkinMul(o: GrumpkinMul): MsgpackGrumpkinMul {
  if (o.point === undefined) { throw new Error("Expected point in GrumpkinMul serialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in GrumpkinMul serialization"); };
  return {
  point: fromGrumpkinPoint(o.point),
  scalar: o.scalar,
  };
}

export function fromGrumpkinAdd(o: GrumpkinAdd): MsgpackGrumpkinAdd {
  if (o.pointA === undefined) { throw new Error("Expected pointA in GrumpkinAdd serialization"); }
  if (o.pointB === undefined) { throw new Error("Expected pointB in GrumpkinAdd serialization"); };
  return {
  point_a: fromGrumpkinPoint(o.pointA),
  point_b: fromGrumpkinPoint(o.pointB),
  };
}

export function fromGrumpkinBatchMul(o: GrumpkinBatchMul): MsgpackGrumpkinBatchMul {
  if (o.points === undefined) { throw new Error("Expected points in GrumpkinBatchMul serialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in GrumpkinBatchMul serialization"); };
  return {
  points: o.points.map((v: any) => fromGrumpkinPoint(v)),
  scalar: o.scalar,
  };
}

export function fromGrumpkinGetRandomFr(o: GrumpkinGetRandomFr): MsgpackGrumpkinGetRandomFr {
  if (o.dummy === undefined) { throw new Error("Expected dummy in GrumpkinGetRandomFr serialization"); };
  return {
  dummy: o.dummy,
  };
}

export function fromGrumpkinReduce512(o: GrumpkinReduce512): MsgpackGrumpkinReduce512 {
  if (o.input === undefined) { throw new Error("Expected input in GrumpkinReduce512 serialization"); };
  return {
  input: o.input,
  };
}

export function fromSecp256k1Point(o: Secp256k1Point): MsgpackSecp256k1Point {
  if (o.x === undefined) { throw new Error("Expected x in Secp256k1Point serialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Secp256k1Point serialization"); };
  return {
  x: o.x,
  y: o.y,
  };
}

export function fromSecp256k1Mul(o: Secp256k1Mul): MsgpackSecp256k1Mul {
  if (o.point === undefined) { throw new Error("Expected point in Secp256k1Mul serialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in Secp256k1Mul serialization"); };
  return {
  point: fromSecp256k1Point(o.point),
  scalar: o.scalar,
  };
}

export function fromSecp256k1GetRandomFr(o: Secp256k1GetRandomFr): MsgpackSecp256k1GetRandomFr {
  if (o.dummy === undefined) { throw new Error("Expected dummy in Secp256k1GetRandomFr serialization"); };
  return {
  dummy: o.dummy,
  };
}

export function fromSecp256k1Reduce512(o: Secp256k1Reduce512): MsgpackSecp256k1Reduce512 {
  if (o.input === undefined) { throw new Error("Expected input in Secp256k1Reduce512 serialization"); };
  return {
  input: o.input,
  };
}

export function fromBn254FrSqrt(o: Bn254FrSqrt): MsgpackBn254FrSqrt {
  if (o.input === undefined) { throw new Error("Expected input in Bn254FrSqrt serialization"); };
  return {
  input: o.input,
  };
}

export function fromBn254FqSqrt(o: Bn254FqSqrt): MsgpackBn254FqSqrt {
  if (o.input === undefined) { throw new Error("Expected input in Bn254FqSqrt serialization"); };
  return {
  input: o.input,
  };
}

export function fromBn254G1Point(o: Bn254G1Point): MsgpackBn254G1Point {
  if (o.x === undefined) { throw new Error("Expected x in Bn254G1Point serialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Bn254G1Point serialization"); };
  return {
  x: o.x,
  y: o.y,
  };
}

export function fromBn254G1Mul(o: Bn254G1Mul): MsgpackBn254G1Mul {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1Mul serialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in Bn254G1Mul serialization"); };
  return {
  point: fromBn254G1Point(o.point),
  scalar: o.scalar,
  };
}

export function fromBn254G2Point(o: Bn254G2Point): MsgpackBn254G2Point {
  if (o.x === undefined) { throw new Error("Expected x in Bn254G2Point serialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Bn254G2Point serialization"); };
  return {
  x: o.x,
  y: o.y,
  };
}

export function fromBn254G2Mul(o: Bn254G2Mul): MsgpackBn254G2Mul {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G2Mul serialization"); }
  if (o.scalar === undefined) { throw new Error("Expected scalar in Bn254G2Mul serialization"); };
  return {
  point: fromBn254G2Point(o.point),
  scalar: o.scalar,
  };
}

export function fromBn254G1IsOnCurve(o: Bn254G1IsOnCurve): MsgpackBn254G1IsOnCurve {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1IsOnCurve serialization"); };
  return {
  point: fromBn254G1Point(o.point),
  };
}

export function fromBn254G1FromCompressed(o: Bn254G1FromCompressed): MsgpackBn254G1FromCompressed {
  if (o.compressed === undefined) { throw new Error("Expected compressed in Bn254G1FromCompressed serialization"); };
  return {
  compressed: o.compressed,
  };
}

export function fromSchnorrComputePublicKey(o: SchnorrComputePublicKey): MsgpackSchnorrComputePublicKey {
  if (o.privateKey === undefined) { throw new Error("Expected privateKey in SchnorrComputePublicKey serialization"); };
  return {
  private_key: o.privateKey,
  };
}

export function fromSchnorrConstructSignature(o: SchnorrConstructSignature): MsgpackSchnorrConstructSignature {
  if (o.message === undefined) { throw new Error("Expected message in SchnorrConstructSignature serialization"); }
  if (o.privateKey === undefined) { throw new Error("Expected privateKey in SchnorrConstructSignature serialization"); };
  return {
  message: o.message,
  private_key: o.privateKey,
  };
}

export function fromSchnorrVerifySignature(o: SchnorrVerifySignature): MsgpackSchnorrVerifySignature {
  if (o.message === undefined) { throw new Error("Expected message in SchnorrVerifySignature serialization"); }
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in SchnorrVerifySignature serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in SchnorrVerifySignature serialization"); }
  if (o.e === undefined) { throw new Error("Expected e in SchnorrVerifySignature serialization"); };
  return {
  message: o.message,
  public_key: fromGrumpkinPoint(o.publicKey),
  s: o.s,
  e: o.e,
  };
}

export function fromEcdsaSecp256k1ComputePublicKey(o: EcdsaSecp256k1ComputePublicKey): MsgpackEcdsaSecp256k1ComputePublicKey {
  if (o.privateKey === undefined) { throw new Error("Expected privateKey in EcdsaSecp256k1ComputePublicKey serialization"); };
  return {
  private_key: o.privateKey,
  };
}

export function fromEcdsaSecp256r1ComputePublicKey(o: EcdsaSecp256r1ComputePublicKey): MsgpackEcdsaSecp256r1ComputePublicKey {
  if (o.privateKey === undefined) { throw new Error("Expected privateKey in EcdsaSecp256r1ComputePublicKey serialization"); };
  return {
  private_key: o.privateKey,
  };
}

export function fromEcdsaSecp256k1ConstructSignature(o: EcdsaSecp256k1ConstructSignature): MsgpackEcdsaSecp256k1ConstructSignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256k1ConstructSignature serialization"); }
  if (o.privateKey === undefined) { throw new Error("Expected privateKey in EcdsaSecp256k1ConstructSignature serialization"); };
  return {
  message: o.message,
  private_key: o.privateKey,
  };
}

export function fromEcdsaSecp256r1ConstructSignature(o: EcdsaSecp256r1ConstructSignature): MsgpackEcdsaSecp256r1ConstructSignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256r1ConstructSignature serialization"); }
  if (o.privateKey === undefined) { throw new Error("Expected privateKey in EcdsaSecp256r1ConstructSignature serialization"); };
  return {
  message: o.message,
  private_key: o.privateKey,
  };
}

export function fromEcdsaSecp256k1RecoverPublicKey(o: EcdsaSecp256k1RecoverPublicKey): MsgpackEcdsaSecp256k1RecoverPublicKey {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256k1RecoverPublicKey serialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256k1RecoverPublicKey serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256k1RecoverPublicKey serialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256k1RecoverPublicKey serialization"); };
  return {
  message: o.message,
  r: o.r,
  s: o.s,
  v: o.v,
  };
}

export function fromEcdsaSecp256r1RecoverPublicKey(o: EcdsaSecp256r1RecoverPublicKey): MsgpackEcdsaSecp256r1RecoverPublicKey {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256r1RecoverPublicKey serialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256r1RecoverPublicKey serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256r1RecoverPublicKey serialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256r1RecoverPublicKey serialization"); };
  return {
  message: o.message,
  r: o.r,
  s: o.s,
  v: o.v,
  };
}

export function fromEcdsaSecp256k1VerifySignature(o: EcdsaSecp256k1VerifySignature): MsgpackEcdsaSecp256k1VerifySignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256k1VerifySignature serialization"); }
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in EcdsaSecp256k1VerifySignature serialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256k1VerifySignature serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256k1VerifySignature serialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256k1VerifySignature serialization"); };
  return {
  message: o.message,
  public_key: fromSecp256k1Point(o.publicKey),
  r: o.r,
  s: o.s,
  v: o.v,
  };
}

export function fromSecp256r1Point(o: Secp256r1Point): MsgpackSecp256r1Point {
  if (o.x === undefined) { throw new Error("Expected x in Secp256r1Point serialization"); }
  if (o.y === undefined) { throw new Error("Expected y in Secp256r1Point serialization"); };
  return {
  x: o.x,
  y: o.y,
  };
}

export function fromEcdsaSecp256r1VerifySignature(o: EcdsaSecp256r1VerifySignature): MsgpackEcdsaSecp256r1VerifySignature {
  if (o.message === undefined) { throw new Error("Expected message in EcdsaSecp256r1VerifySignature serialization"); }
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in EcdsaSecp256r1VerifySignature serialization"); }
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256r1VerifySignature serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256r1VerifySignature serialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256r1VerifySignature serialization"); };
  return {
  message: o.message,
  public_key: fromSecp256r1Point(o.publicKey),
  r: o.r,
  s: o.s,
  v: o.v,
  };
}

export function fromSrsInitSrs(o: SrsInitSrs): MsgpackSrsInitSrs {
  if (o.pointsBuf === undefined) { throw new Error("Expected pointsBuf in SrsInitSrs serialization"); }
  if (o.numPoints === undefined) { throw new Error("Expected numPoints in SrsInitSrs serialization"); }
  if (o.g2Point === undefined) { throw new Error("Expected g2Point in SrsInitSrs serialization"); };
  return {
  points_buf: o.pointsBuf,
  num_points: o.numPoints,
  g2_point: o.g2Point,
  };
}

export function fromSrsInitGrumpkinSrs(o: SrsInitGrumpkinSrs): MsgpackSrsInitGrumpkinSrs {
  if (o.pointsBuf === undefined) { throw new Error("Expected pointsBuf in SrsInitGrumpkinSrs serialization"); }
  if (o.numPoints === undefined) { throw new Error("Expected numPoints in SrsInitGrumpkinSrs serialization"); };
  return {
  points_buf: o.pointsBuf,
  num_points: o.numPoints,
  };
}

export function fromShutdown(o: Shutdown): MsgpackShutdown {
  return {};
}

export function fromErrorResponse(o: ErrorResponse): MsgpackErrorResponse {
  if (o.message === undefined) { throw new Error("Expected message in ErrorResponse serialization"); };
  return {
  message: o.message,
  };
}

export function fromCircuitProveResponse(o: CircuitProveResponse): MsgpackCircuitProveResponse {
  if (o.publicInputs === undefined) { throw new Error("Expected publicInputs in CircuitProveResponse serialization"); }
  if (o.proof === undefined) { throw new Error("Expected proof in CircuitProveResponse serialization"); }
  if (o.vk === undefined) { throw new Error("Expected vk in CircuitProveResponse serialization"); };
  return {
  public_inputs: o.publicInputs,
  proof: o.proof,
  vk: fromCircuitComputeVkResponse(o.vk),
  };
}

export function fromCircuitInfoResponse(o: CircuitInfoResponse): MsgpackCircuitInfoResponse {
  if (o.numGates === undefined) { throw new Error("Expected numGates in CircuitInfoResponse serialization"); }
  if (o.numGatesDyadic === undefined) { throw new Error("Expected numGatesDyadic in CircuitInfoResponse serialization"); }
  if (o.numAcirOpcodes === undefined) { throw new Error("Expected numAcirOpcodes in CircuitInfoResponse serialization"); }
  if (o.gatesPerOpcode === undefined) { throw new Error("Expected gatesPerOpcode in CircuitInfoResponse serialization"); };
  return {
  num_gates: o.numGates,
  num_gates_dyadic: o.numGatesDyadic,
  num_acir_opcodes: o.numAcirOpcodes,
  gates_per_opcode: o.gatesPerOpcode,
  };
}

export function fromCircuitVerifyResponse(o: CircuitVerifyResponse): MsgpackCircuitVerifyResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in CircuitVerifyResponse serialization"); };
  return {
  verified: o.verified,
  };
}

export function fromChonkComputeStandaloneVkResponse(o: ChonkComputeStandaloneVkResponse): MsgpackChonkComputeStandaloneVkResponse {
  if (o.bytes === undefined) { throw new Error("Expected bytes in ChonkComputeStandaloneVkResponse serialization"); }
  if (o.fields === undefined) { throw new Error("Expected fields in ChonkComputeStandaloneVkResponse serialization"); };
  return {
  bytes: o.bytes,
  fields: o.fields,
  };
}

export function fromChonkComputeIvcVkResponse(o: ChonkComputeIvcVkResponse): MsgpackChonkComputeIvcVkResponse {
  if (o.bytes === undefined) { throw new Error("Expected bytes in ChonkComputeIvcVkResponse serialization"); };
  return {
  bytes: o.bytes,
  };
}

export function fromChonkStartResponse(o: ChonkStartResponse): MsgpackChonkStartResponse {
  return {};
}

export function fromChonkLoadResponse(o: ChonkLoadResponse): MsgpackChonkLoadResponse {
  return {};
}

export function fromChonkAccumulateResponse(o: ChonkAccumulateResponse): MsgpackChonkAccumulateResponse {
  return {};
}

export function fromChonkProveResponse(o: ChonkProveResponse): MsgpackChonkProveResponse {
  if (o.proof === undefined) { throw new Error("Expected proof in ChonkProveResponse serialization"); };
  return {
  proof: fromChonkProof(o.proof),
  };
}

export function fromChonkVerifyResponse(o: ChonkVerifyResponse): MsgpackChonkVerifyResponse {
  if (o.valid === undefined) { throw new Error("Expected valid in ChonkVerifyResponse serialization"); };
  return {
  valid: o.valid,
  };
}

export function fromVkAsFieldsResponse(o: VkAsFieldsResponse): MsgpackVkAsFieldsResponse {
  if (o.fields === undefined) { throw new Error("Expected fields in VkAsFieldsResponse serialization"); };
  return {
  fields: o.fields,
  };
}

export function fromMegaVkAsFieldsResponse(o: MegaVkAsFieldsResponse): MsgpackMegaVkAsFieldsResponse {
  if (o.fields === undefined) { throw new Error("Expected fields in MegaVkAsFieldsResponse serialization"); };
  return {
  fields: o.fields,
  };
}

export function fromCircuitWriteSolidityVerifierResponse(o: CircuitWriteSolidityVerifierResponse): MsgpackCircuitWriteSolidityVerifierResponse {
  if (o.solidityCode === undefined) { throw new Error("Expected solidityCode in CircuitWriteSolidityVerifierResponse serialization"); };
  return {
  solidity_code: o.solidityCode,
  };
}

export function fromChonkCheckPrecomputedVkResponse(o: ChonkCheckPrecomputedVkResponse): MsgpackChonkCheckPrecomputedVkResponse {
  if (o.valid === undefined) { throw new Error("Expected valid in ChonkCheckPrecomputedVkResponse serialization"); }
  if (o.actualVk === undefined) { throw new Error("Expected actualVk in ChonkCheckPrecomputedVkResponse serialization"); };
  return {
  valid: o.valid,
  actual_vk: o.actualVk,
  };
}

export function fromChonkStatsResponse(o: ChonkStatsResponse): MsgpackChonkStatsResponse {
  if (o.acirOpcodes === undefined) { throw new Error("Expected acirOpcodes in ChonkStatsResponse serialization"); }
  if (o.circuitSize === undefined) { throw new Error("Expected circuitSize in ChonkStatsResponse serialization"); }
  if (o.gatesPerOpcode === undefined) { throw new Error("Expected gatesPerOpcode in ChonkStatsResponse serialization"); };
  return {
  acir_opcodes: o.acirOpcodes,
  circuit_size: o.circuitSize,
  gates_per_opcode: o.gatesPerOpcode,
  };
}

export function fromPoseidon2HashResponse(o: Poseidon2HashResponse): MsgpackPoseidon2HashResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in Poseidon2HashResponse serialization"); };
  return {
  hash: o.hash,
  };
}

export function fromPoseidon2PermutationResponse(o: Poseidon2PermutationResponse): MsgpackPoseidon2PermutationResponse {
  if (o.outputs === undefined) { throw new Error("Expected outputs in Poseidon2PermutationResponse serialization"); };
  return {
  outputs: o.outputs,
  };
}

export function fromPoseidon2HashAccumulateResponse(o: Poseidon2HashAccumulateResponse): MsgpackPoseidon2HashAccumulateResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in Poseidon2HashAccumulateResponse serialization"); };
  return {
  hash: o.hash,
  };
}

export function fromPedersenCommitResponse(o: PedersenCommitResponse): MsgpackPedersenCommitResponse {
  if (o.point === undefined) { throw new Error("Expected point in PedersenCommitResponse serialization"); };
  return {
  point: fromGrumpkinPoint(o.point),
  };
}

export function fromPedersenHashResponse(o: PedersenHashResponse): MsgpackPedersenHashResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in PedersenHashResponse serialization"); };
  return {
  hash: o.hash,
  };
}

export function fromPedersenHashBufferResponse(o: PedersenHashBufferResponse): MsgpackPedersenHashBufferResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in PedersenHashBufferResponse serialization"); };
  return {
  hash: o.hash,
  };
}

export function fromBlake2sResponse(o: Blake2sResponse): MsgpackBlake2sResponse {
  if (o.hash === undefined) { throw new Error("Expected hash in Blake2sResponse serialization"); };
  return {
  hash: o.hash,
  };
}

export function fromBlake2sToFieldResponse(o: Blake2sToFieldResponse): MsgpackBlake2sToFieldResponse {
  if (o.field === undefined) { throw new Error("Expected field in Blake2sToFieldResponse serialization"); };
  return {
  field: o.field,
  };
}

export function fromAesEncryptResponse(o: AesEncryptResponse): MsgpackAesEncryptResponse {
  if (o.ciphertext === undefined) { throw new Error("Expected ciphertext in AesEncryptResponse serialization"); };
  return {
  ciphertext: o.ciphertext,
  };
}

export function fromAesDecryptResponse(o: AesDecryptResponse): MsgpackAesDecryptResponse {
  if (o.plaintext === undefined) { throw new Error("Expected plaintext in AesDecryptResponse serialization"); };
  return {
  plaintext: o.plaintext,
  };
}

export function fromGrumpkinMulResponse(o: GrumpkinMulResponse): MsgpackGrumpkinMulResponse {
  if (o.point === undefined) { throw new Error("Expected point in GrumpkinMulResponse serialization"); };
  return {
  point: fromGrumpkinPoint(o.point),
  };
}

export function fromGrumpkinAddResponse(o: GrumpkinAddResponse): MsgpackGrumpkinAddResponse {
  if (o.point === undefined) { throw new Error("Expected point in GrumpkinAddResponse serialization"); };
  return {
  point: fromGrumpkinPoint(o.point),
  };
}

export function fromGrumpkinBatchMulResponse(o: GrumpkinBatchMulResponse): MsgpackGrumpkinBatchMulResponse {
  if (o.points === undefined) { throw new Error("Expected points in GrumpkinBatchMulResponse serialization"); };
  return {
  points: o.points.map((v: any) => fromGrumpkinPoint(v)),
  };
}

export function fromGrumpkinGetRandomFrResponse(o: GrumpkinGetRandomFrResponse): MsgpackGrumpkinGetRandomFrResponse {
  if (o.value === undefined) { throw new Error("Expected value in GrumpkinGetRandomFrResponse serialization"); };
  return {
  value: o.value,
  };
}

export function fromGrumpkinReduce512Response(o: GrumpkinReduce512Response): MsgpackGrumpkinReduce512Response {
  if (o.value === undefined) { throw new Error("Expected value in GrumpkinReduce512Response serialization"); };
  return {
  value: o.value,
  };
}

export function fromSecp256k1MulResponse(o: Secp256k1MulResponse): MsgpackSecp256k1MulResponse {
  if (o.point === undefined) { throw new Error("Expected point in Secp256k1MulResponse serialization"); };
  return {
  point: fromSecp256k1Point(o.point),
  };
}

export function fromSecp256k1GetRandomFrResponse(o: Secp256k1GetRandomFrResponse): MsgpackSecp256k1GetRandomFrResponse {
  if (o.value === undefined) { throw new Error("Expected value in Secp256k1GetRandomFrResponse serialization"); };
  return {
  value: o.value,
  };
}

export function fromSecp256k1Reduce512Response(o: Secp256k1Reduce512Response): MsgpackSecp256k1Reduce512Response {
  if (o.value === undefined) { throw new Error("Expected value in Secp256k1Reduce512Response serialization"); };
  return {
  value: o.value,
  };
}

export function fromBn254FrSqrtResponse(o: Bn254FrSqrtResponse): MsgpackBn254FrSqrtResponse {
  if (o.isSquareRoot === undefined) { throw new Error("Expected isSquareRoot in Bn254FrSqrtResponse serialization"); }
  if (o.value === undefined) { throw new Error("Expected value in Bn254FrSqrtResponse serialization"); };
  return {
  is_square_root: o.isSquareRoot,
  value: o.value,
  };
}

export function fromBn254FqSqrtResponse(o: Bn254FqSqrtResponse): MsgpackBn254FqSqrtResponse {
  if (o.isSquareRoot === undefined) { throw new Error("Expected isSquareRoot in Bn254FqSqrtResponse serialization"); }
  if (o.value === undefined) { throw new Error("Expected value in Bn254FqSqrtResponse serialization"); };
  return {
  is_square_root: o.isSquareRoot,
  value: o.value,
  };
}

export function fromBn254G1MulResponse(o: Bn254G1MulResponse): MsgpackBn254G1MulResponse {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1MulResponse serialization"); };
  return {
  point: fromBn254G1Point(o.point),
  };
}

export function fromBn254G2MulResponse(o: Bn254G2MulResponse): MsgpackBn254G2MulResponse {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G2MulResponse serialization"); };
  return {
  point: fromBn254G2Point(o.point),
  };
}

export function fromBn254G1IsOnCurveResponse(o: Bn254G1IsOnCurveResponse): MsgpackBn254G1IsOnCurveResponse {
  if (o.isOnCurve === undefined) { throw new Error("Expected isOnCurve in Bn254G1IsOnCurveResponse serialization"); };
  return {
  is_on_curve: o.isOnCurve,
  };
}

export function fromBn254G1FromCompressedResponse(o: Bn254G1FromCompressedResponse): MsgpackBn254G1FromCompressedResponse {
  if (o.point === undefined) { throw new Error("Expected point in Bn254G1FromCompressedResponse serialization"); };
  return {
  point: fromBn254G1Point(o.point),
  };
}

export function fromSchnorrComputePublicKeyResponse(o: SchnorrComputePublicKeyResponse): MsgpackSchnorrComputePublicKeyResponse {
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in SchnorrComputePublicKeyResponse serialization"); };
  return {
  public_key: fromGrumpkinPoint(o.publicKey),
  };
}

export function fromSchnorrConstructSignatureResponse(o: SchnorrConstructSignatureResponse): MsgpackSchnorrConstructSignatureResponse {
  if (o.s === undefined) { throw new Error("Expected s in SchnorrConstructSignatureResponse serialization"); }
  if (o.e === undefined) { throw new Error("Expected e in SchnorrConstructSignatureResponse serialization"); };
  return {
  s: o.s,
  e: o.e,
  };
}

export function fromSchnorrVerifySignatureResponse(o: SchnorrVerifySignatureResponse): MsgpackSchnorrVerifySignatureResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in SchnorrVerifySignatureResponse serialization"); };
  return {
  verified: o.verified,
  };
}

export function fromEcdsaSecp256k1ComputePublicKeyResponse(o: EcdsaSecp256k1ComputePublicKeyResponse): MsgpackEcdsaSecp256k1ComputePublicKeyResponse {
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in EcdsaSecp256k1ComputePublicKeyResponse serialization"); };
  return {
  public_key: fromSecp256k1Point(o.publicKey),
  };
}

export function fromEcdsaSecp256r1ComputePublicKeyResponse(o: EcdsaSecp256r1ComputePublicKeyResponse): MsgpackEcdsaSecp256r1ComputePublicKeyResponse {
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in EcdsaSecp256r1ComputePublicKeyResponse serialization"); };
  return {
  public_key: fromSecp256r1Point(o.publicKey),
  };
}

export function fromEcdsaSecp256k1ConstructSignatureResponse(o: EcdsaSecp256k1ConstructSignatureResponse): MsgpackEcdsaSecp256k1ConstructSignatureResponse {
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256k1ConstructSignatureResponse serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256k1ConstructSignatureResponse serialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256k1ConstructSignatureResponse serialization"); };
  return {
  r: o.r,
  s: o.s,
  v: o.v,
  };
}

export function fromEcdsaSecp256r1ConstructSignatureResponse(o: EcdsaSecp256r1ConstructSignatureResponse): MsgpackEcdsaSecp256r1ConstructSignatureResponse {
  if (o.r === undefined) { throw new Error("Expected r in EcdsaSecp256r1ConstructSignatureResponse serialization"); }
  if (o.s === undefined) { throw new Error("Expected s in EcdsaSecp256r1ConstructSignatureResponse serialization"); }
  if (o.v === undefined) { throw new Error("Expected v in EcdsaSecp256r1ConstructSignatureResponse serialization"); };
  return {
  r: o.r,
  s: o.s,
  v: o.v,
  };
}

export function fromEcdsaSecp256k1RecoverPublicKeyResponse(o: EcdsaSecp256k1RecoverPublicKeyResponse): MsgpackEcdsaSecp256k1RecoverPublicKeyResponse {
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in EcdsaSecp256k1RecoverPublicKeyResponse serialization"); };
  return {
  public_key: fromSecp256k1Point(o.publicKey),
  };
}

export function fromEcdsaSecp256r1RecoverPublicKeyResponse(o: EcdsaSecp256r1RecoverPublicKeyResponse): MsgpackEcdsaSecp256r1RecoverPublicKeyResponse {
  if (o.publicKey === undefined) { throw new Error("Expected publicKey in EcdsaSecp256r1RecoverPublicKeyResponse serialization"); };
  return {
  public_key: fromSecp256r1Point(o.publicKey),
  };
}

export function fromEcdsaSecp256k1VerifySignatureResponse(o: EcdsaSecp256k1VerifySignatureResponse): MsgpackEcdsaSecp256k1VerifySignatureResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in EcdsaSecp256k1VerifySignatureResponse serialization"); };
  return {
  verified: o.verified,
  };
}

export function fromEcdsaSecp256r1VerifySignatureResponse(o: EcdsaSecp256r1VerifySignatureResponse): MsgpackEcdsaSecp256r1VerifySignatureResponse {
  if (o.verified === undefined) { throw new Error("Expected verified in EcdsaSecp256r1VerifySignatureResponse serialization"); };
  return {
  verified: o.verified,
  };
}

export function fromSrsInitSrsResponse(o: SrsInitSrsResponse): MsgpackSrsInitSrsResponse {
  if (o.dummy === undefined) { throw new Error("Expected dummy in SrsInitSrsResponse serialization"); };
  return {
  dummy: o.dummy,
  };
}

export function fromSrsInitGrumpkinSrsResponse(o: SrsInitGrumpkinSrsResponse): MsgpackSrsInitGrumpkinSrsResponse {
  if (o.dummy === undefined) { throw new Error("Expected dummy in SrsInitGrumpkinSrsResponse serialization"); };
  return {
  dummy: o.dummy,
  };
}

export function fromShutdownResponse(o: ShutdownResponse): MsgpackShutdownResponse {
  return {};
}

// Base API interface
export interface BbApiBase {
  circuitProve(command: CircuitProve): Promise<CircuitProveResponse>;
  circuitComputeVk(command: CircuitComputeVk): Promise<CircuitComputeVkResponse>;
  circuitStats(command: CircuitStats): Promise<CircuitInfoResponse>;
  circuitVerify(command: CircuitVerify): Promise<CircuitVerifyResponse>;
  chonkComputeStandaloneVk(command: ChonkComputeStandaloneVk): Promise<ChonkComputeStandaloneVkResponse>;
  chonkComputeIvcVk(command: ChonkComputeIvcVk): Promise<ChonkComputeIvcVkResponse>;
  chonkStart(command: ChonkStart): Promise<ChonkStartResponse>;
  chonkLoad(command: ChonkLoad): Promise<ChonkLoadResponse>;
  chonkAccumulate(command: ChonkAccumulate): Promise<ChonkAccumulateResponse>;
  chonkProve(command: ChonkProve): Promise<ChonkProveResponse>;
  chonkVerify(command: ChonkVerify): Promise<ChonkVerifyResponse>;
  vkAsFields(command: VkAsFields): Promise<VkAsFieldsResponse>;
  megaVkAsFields(command: MegaVkAsFields): Promise<MegaVkAsFieldsResponse>;
  circuitWriteSolidityVerifier(command: CircuitWriteSolidityVerifier): Promise<CircuitWriteSolidityVerifierResponse>;
  chonkCheckPrecomputedVk(command: ChonkCheckPrecomputedVk): Promise<ChonkCheckPrecomputedVkResponse>;
  chonkStats(command: ChonkStats): Promise<ChonkStatsResponse>;
  poseidon2Hash(command: Poseidon2Hash): Promise<Poseidon2HashResponse>;
  poseidon2Permutation(command: Poseidon2Permutation): Promise<Poseidon2PermutationResponse>;
  poseidon2HashAccumulate(command: Poseidon2HashAccumulate): Promise<Poseidon2HashAccumulateResponse>;
  pedersenCommit(command: PedersenCommit): Promise<PedersenCommitResponse>;
  pedersenHash(command: PedersenHash): Promise<PedersenHashResponse>;
  pedersenHashBuffer(command: PedersenHashBuffer): Promise<PedersenHashBufferResponse>;
  blake2s(command: Blake2s): Promise<Blake2sResponse>;
  blake2sToField(command: Blake2sToField): Promise<Blake2sToFieldResponse>;
  aesEncrypt(command: AesEncrypt): Promise<AesEncryptResponse>;
  aesDecrypt(command: AesDecrypt): Promise<AesDecryptResponse>;
  grumpkinMul(command: GrumpkinMul): Promise<GrumpkinMulResponse>;
  grumpkinAdd(command: GrumpkinAdd): Promise<GrumpkinAddResponse>;
  grumpkinBatchMul(command: GrumpkinBatchMul): Promise<GrumpkinBatchMulResponse>;
  grumpkinGetRandomFr(command: GrumpkinGetRandomFr): Promise<GrumpkinGetRandomFrResponse>;
  grumpkinReduce512(command: GrumpkinReduce512): Promise<GrumpkinReduce512Response>;
  secp256k1Mul(command: Secp256k1Mul): Promise<Secp256k1MulResponse>;
  secp256k1GetRandomFr(command: Secp256k1GetRandomFr): Promise<Secp256k1GetRandomFrResponse>;
  secp256k1Reduce512(command: Secp256k1Reduce512): Promise<Secp256k1Reduce512Response>;
  bn254FrSqrt(command: Bn254FrSqrt): Promise<Bn254FrSqrtResponse>;
  bn254FqSqrt(command: Bn254FqSqrt): Promise<Bn254FqSqrtResponse>;
  bn254G1Mul(command: Bn254G1Mul): Promise<Bn254G1MulResponse>;
  bn254G2Mul(command: Bn254G2Mul): Promise<Bn254G2MulResponse>;
  bn254G1IsOnCurve(command: Bn254G1IsOnCurve): Promise<Bn254G1IsOnCurveResponse>;
  bn254G1FromCompressed(command: Bn254G1FromCompressed): Promise<Bn254G1FromCompressedResponse>;
  schnorrComputePublicKey(command: SchnorrComputePublicKey): Promise<SchnorrComputePublicKeyResponse>;
  schnorrConstructSignature(command: SchnorrConstructSignature): Promise<SchnorrConstructSignatureResponse>;
  schnorrVerifySignature(command: SchnorrVerifySignature): Promise<SchnorrVerifySignatureResponse>;
  ecdsaSecp256k1ComputePublicKey(command: EcdsaSecp256k1ComputePublicKey): Promise<EcdsaSecp256k1ComputePublicKeyResponse>;
  ecdsaSecp256r1ComputePublicKey(command: EcdsaSecp256r1ComputePublicKey): Promise<EcdsaSecp256r1ComputePublicKeyResponse>;
  ecdsaSecp256k1ConstructSignature(command: EcdsaSecp256k1ConstructSignature): Promise<EcdsaSecp256k1ConstructSignatureResponse>;
  ecdsaSecp256r1ConstructSignature(command: EcdsaSecp256r1ConstructSignature): Promise<EcdsaSecp256r1ConstructSignatureResponse>;
  ecdsaSecp256k1RecoverPublicKey(command: EcdsaSecp256k1RecoverPublicKey): Promise<EcdsaSecp256k1RecoverPublicKeyResponse>;
  ecdsaSecp256r1RecoverPublicKey(command: EcdsaSecp256r1RecoverPublicKey): Promise<EcdsaSecp256r1RecoverPublicKeyResponse>;
  ecdsaSecp256k1VerifySignature(command: EcdsaSecp256k1VerifySignature): Promise<EcdsaSecp256k1VerifySignatureResponse>;
  ecdsaSecp256r1VerifySignature(command: EcdsaSecp256r1VerifySignature): Promise<EcdsaSecp256r1VerifySignatureResponse>;
  srsInitSrs(command: SrsInitSrs): Promise<SrsInitSrsResponse>;
  srsInitGrumpkinSrs(command: SrsInitGrumpkinSrs): Promise<SrsInitGrumpkinSrsResponse>;
  shutdown(command: Shutdown): Promise<ShutdownResponse>;
  destroy(): Promise<void>;
}
